{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<a id='io'></a>\n",
    "{{ header }}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# IO tools (text, CSV, HDF5, …)\n",
    "\n",
    "The pandas I/O API is a set of top level `reader` functions accessed like\n",
    "`pandas.read_csv()` that generally return a pandas object. The corresponding\n",
    "`writer` functions are object methods that are accessed like\n",
    "`DataFrame.to_csv()`. Below is a table containing available `readers` and\n",
    "`writers`.\n",
    "\n",
    "[](https://en.wikipedia.org/wiki/Comma-separated_values)[](#io-read-csv-table)[](#io-store-in-csv)[](https://www.json.org/)[](#io-json-reader)[](#io-json-writer)[](https://en.wikipedia.org/wiki/HTML)[](#io-read-html)[](#io-html)[](#io-clipboard)[](#io-clipboard)[](https://en.wikipedia.org/wiki/Microsoft_Excel)[](#io-excel-reader)[](#io-excel-writer)[](http://www.opendocumentformat.org)[](#io-ods)[](https://support.hdfgroup.org/HDF5/whatishdf5.html)[](#io-hdf5)[](#io-hdf5)[](https://github.com/wesm/feather)[](#io-feather)[](#io-feather)[](https://parquet.apache.org/)[](#io-parquet)[](#io-parquet)[](https://msgpack.org/index.html)[](#io-msgpack)[](#io-msgpack)[](https://en.wikipedia.org/wiki/Stata)[](#io-stata-reader)[](#io-stata-writer)[](https://en.wikipedia.org/wiki/SAS_%28software%29)[](#io-sas-reader)[](https://docs.python.org/3/library/pickle.html)[](#io-pickle)[](#io-pickle)[](https://en.wikipedia.org/wiki/SQL)[](#io-sql)[](#io-sql)[](https://en.wikipedia.org/wiki/BigQuery)[](#io-bigquery)[](#io-bigquery)|Format Type|Data Description|Reader|Writer|\n",
    "|:----------------------------:|:--------------------------------------------------------------------------------------------------:|:----------------------------------------------------------:|:----------------------------------------------------------:|\n",
    "|text|CSV|read_csv|to_csv|\n",
    "|text|JSON|read_json|to_json|\n",
    "|text|HTML|read_html|to_html|\n",
    "|text|Local clipboard|read_clipboard|to_clipboard|\n",
    "|binary|MS Excel|read_excel|to_excel|\n",
    "|binary|OpenDocument|read_excel||\n",
    "|binary|HDF5 Format|read_hdf|to_hdf|\n",
    "|binary|Feather Format|read_feather|to_feather|\n",
    "|binary|Parquet Format|read_parquet|to_parquet|\n",
    "|binary|Msgpack|read_msgpack|to_msgpack|\n",
    "|binary|Stata|read_stata|to_stata|\n",
    "|binary|SAS|read_sas||\n",
    "|binary|Python Pickle Format|read_pickle|to_pickle|\n",
    "|SQL|SQL|read_sql|to_sql|\n",
    "|SQL|Google Big Query|read_gbq|to_gbq|\n",
    "[Here](#io-perf) is an informal performance comparison for some of these IO methods.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">For examples that use the `StringIO` class, make sure you import it\n",
    "according to your Python version, i.e. `from StringIO import StringIO` for\n",
    "Python 2 and `from io import StringIO` for Python 3.\n",
    "\n",
    "\n",
    "<a id='io-read-csv-table'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CSV & text files\n",
    "\n",
    "The workhorse function for reading text files (a.k.a. flat files) is\n",
    "`read_csv()`. See the [cookbook](cookbook.ipynb#cookbook-csv) for some advanced strategies."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parsing options\n",
    "\n",
    "`read_csv()` accepts the following common arguments:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Basic\n",
    "\n",
    "\n",
    "<dl style='margin: 20px 0;'>\n",
    "<dt>filepath_or_buffer</dt>\n",
    "various<dd>\n",
    "Either a path to a file (a [`str`](https://docs.python.org/3/library/stdtypes.html#str), [`pathlib.Path`](https://docs.python.org/3/library/pathlib.html#pathlib.Path),\n",
    "or `py._path.local.LocalPath`), URL (including http, ftp, and S3\n",
    "locations), or any object with a `read()` method (such as an open file or\n",
    "[`StringIO`](https://docs.python.org/3/library/io.html#io.StringIO)).\n",
    "\n",
    "</dd>\n",
    "<dt>sep</dt>\n",
    "str, defaults to `','` for `read_csv()`, `\\t` for `read_table()`<dd>\n",
    "Delimiter to use. If sep is `None`, the C engine cannot automatically detect\n",
    "the separator, but the Python parsing engine can, meaning the latter will be\n",
    "used and automatically detect the separator by Python’s builtin sniffer tool,\n",
    "[`csv.Sniffer`](https://docs.python.org/3/library/csv.html#csv.Sniffer). In addition, separators longer than 1 character and\n",
    "different from `'\\s+'` will be interpreted as regular expressions and\n",
    "will also force the use of the Python parsing engine. Note that regex\n",
    "delimiters are prone to ignoring quoted data. Regex example: `'\\\\r\\\\t'`.\n",
    "\n",
    "</dd>\n",
    "<dt>delimiter</dt>\n",
    "str, default `None`<dd>\n",
    "Alternative argument name for sep.\n",
    "\n",
    "</dd>\n",
    "<dt>delim_whitespace</dt>\n",
    "boolean, default False<dd>\n",
    "Specifies whether or not whitespace (e.g. `' '` or `'\\t'`)\n",
    "will be used as the delimiter. Equivalent to setting `sep='\\s+'`.\n",
    "If this option is set to `True`, nothing should be passed in for the\n",
    "`delimiter` parameter.\n",
    "\n",
    "New in version 0.18.1: support for the Python parser.\n",
    "\n",
    "</dd>\n",
    "\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Column and index locations and names\n",
    "\n",
    "\n",
    "<dl style='margin: 20px 0;'>\n",
    "<dt>header</dt>\n",
    "int or list of ints, default `'infer'`<dd>\n",
    "Row number(s) to use as the column names, and the start of the\n",
    "data. Default behavior is to infer the column names: if no names are\n",
    "passed the behavior is identical to `header=0` and column names\n",
    "are inferred from the first line of the file, if column names are\n",
    "passed explicitly then the behavior is identical to\n",
    "`header=None`. Explicitly pass `header=0` to be able to replace\n",
    "existing names.\n",
    "\n",
    "The header can be a list of ints that specify row locations\n",
    "for a MultiIndex on the columns e.g. `[0,1,3]`. Intervening rows\n",
    "that are not specified will be skipped (e.g. 2 in this example is\n",
    "skipped). Note that this parameter ignores commented lines and empty\n",
    "lines if `skip_blank_lines=True`, so header=0 denotes the first\n",
    "line of data rather than the first line of the file.\n",
    "\n",
    "</dd>\n",
    "<dt>names</dt>\n",
    "array-like, default `None`<dd>\n",
    "List of column names to use. If file contains no header row, then you should\n",
    "explicitly pass `header=None`. Duplicates in this list are not allowed.\n",
    "\n",
    "</dd>\n",
    "<dt>index_col</dt>\n",
    "int, str, sequence of int / str, or False, default `None`<dd>\n",
    "Column(s) to use as the row labels of the `DataFrame`, either given as\n",
    "string name or column index. If a sequence of int / str is given, a\n",
    "MultiIndex is used.\n",
    "\n",
    "Note: `index_col=False` can be used to force pandas to *not* use the first\n",
    "column as the index, e.g. when you have a malformed file with delimiters at\n",
    "the end of each line.\n",
    "\n",
    "</dd>\n",
    "<dt>usecols</dt>\n",
    "list-like or callable, default `None`<dd>\n",
    "Return a subset of the columns. If list-like, all elements must either\n",
    "be positional (i.e. integer indices into the document columns) or strings\n",
    "that correspond to column names provided either by the user in names or\n",
    "inferred from the document header row(s). For example, a valid list-like\n",
    "usecols parameter would be `[0, 1, 2]` or `['foo', 'bar', 'baz']`.\n",
    "\n",
    "Element order is ignored, so `usecols=[0, 1]` is the same as `[1, 0]`. To\n",
    "instantiate a DataFrame from `data` with element order preserved use\n",
    "`pd.read_csv(data, usecols=['foo', 'bar'])[['foo', 'bar']]` for columns\n",
    "in `['foo', 'bar']` order or\n",
    "`pd.read_csv(data, usecols=['foo', 'bar'])[['bar', 'foo']]` for\n",
    "`['bar', 'foo']` order.\n",
    "\n",
    "If callable, the callable function will be evaluated against the column names,\n",
    "returning names where the callable function evaluates to True:\n",
    "\n",
    "Using this parameter results in much faster parsing time and lower memory usage.\n",
    "\n",
    "</dd>\n",
    "<dt>squeeze</dt>\n",
    "boolean, default `False`<dd>\n",
    "If the parsed data only contains one column then return a `Series`.\n",
    "\n",
    "</dd>\n",
    "<dt>prefix</dt>\n",
    "str, default `None`<dd>\n",
    "Prefix to add to column numbers when no header, e.g. ‘X’ for X0, X1, …\n",
    "\n",
    "</dd>\n",
    "<dt>mangle_dupe_cols</dt>\n",
    "boolean, default `True`<dd>\n",
    "Duplicate columns will be specified as ‘X’, ‘X.1’…’X.N’, rather than ‘X’…’X’.\n",
    "Passing in `False` will cause data to be overwritten if there are duplicate\n",
    "names in the columns.\n",
    "\n",
    "</dd>\n",
    "\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### General parsing configuration\n",
    "\n",
    "\n",
    "<dl style='margin: 20px 0;'>\n",
    "<dt>dtype</dt>\n",
    "Type name or dict of column -> type, default `None`<dd>\n",
    "Data type for data or columns. E.g. `{'a': np.float64, 'b': np.int32}`\n",
    "(unsupported with `engine='python'`). Use str or object together\n",
    "with suitable `na_values` settings to preserve and\n",
    "not interpret dtype.\n",
    "\n",
    "New in version 0.20.0: support for the Python parser.\n",
    "\n",
    "</dd>\n",
    "<dt>engine</dt>\n",
    "{`'c'`, `'python'`}<dd>\n",
    "Parser engine to use. The C engine is faster while the Python engine is\n",
    "currently more feature-complete.\n",
    "\n",
    "</dd>\n",
    "<dt>converters</dt>\n",
    "dict, default `None`<dd>\n",
    "Dict of functions for converting values in certain columns. Keys can either be\n",
    "integers or column labels.\n",
    "\n",
    "</dd>\n",
    "<dt>true_values</dt>\n",
    "list, default `None`<dd>\n",
    "Values to consider as `True`.\n",
    "\n",
    "</dd>\n",
    "<dt>false_values</dt>\n",
    "list, default `None`<dd>\n",
    "Values to consider as `False`.\n",
    "\n",
    "</dd>\n",
    "<dt>skipinitialspace</dt>\n",
    "boolean, default `False`<dd>\n",
    "Skip spaces after delimiter.\n",
    "\n",
    "</dd>\n",
    "<dt>skiprows</dt>\n",
    "list-like or integer, default `None`<dd>\n",
    "Line numbers to skip (0-indexed) or number of lines to skip (int) at the start\n",
    "of the file.\n",
    "\n",
    "If callable, the callable function will be evaluated against the row\n",
    "indices, returning True if the row should be skipped and False otherwise:\n",
    "\n",
    "</dd>\n",
    "<dt>skipfooter</dt>\n",
    "int, default `0`<dd>\n",
    "Number of lines at bottom of file to skip (unsupported with engine=’c’).\n",
    "\n",
    "</dd>\n",
    "<dt>nrows</dt>\n",
    "int, default `None`<dd>\n",
    "Number of rows of file to read. Useful for reading pieces of large files.\n",
    "\n",
    "</dd>\n",
    "<dt>low_memory</dt>\n",
    "boolean, default `True`<dd>\n",
    "Internally process the file in chunks, resulting in lower memory use\n",
    "while parsing, but possibly mixed type inference.  To ensure no mixed\n",
    "types either set `False`, or specify the type with the `dtype` parameter.\n",
    "Note that the entire file is read into a single `DataFrame` regardless,\n",
    "use the `chunksize` or `iterator` parameter to return the data in chunks.\n",
    "(Only valid with C parser)\n",
    "\n",
    "</dd>\n",
    "<dt>memory_map</dt>\n",
    "boolean, default False<dd>\n",
    "If a filepath is provided for `filepath_or_buffer`, map the file object\n",
    "directly onto memory and access the data directly from there. Using this\n",
    "option can improve performance because there is no longer any I/O overhead.\n",
    "\n",
    "</dd>\n",
    "\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### NA and missing data handling\n",
    "\n",
    "\n",
    "<dl style='margin: 20px 0;'>\n",
    "<dt>na_values</dt>\n",
    "scalar, str, list-like, or dict, default `None`<dd>\n",
    "Additional strings to recognize as NA/NaN. If dict passed, specific per-column\n",
    "NA values. See [na values const](#io-navaluesconst) below\n",
    "for a list of the values interpreted as NaN by default.\n",
    "\n",
    "</dd>\n",
    "<dt>keep_default_na</dt>\n",
    "boolean, default `True`<dd>\n",
    "Whether or not to include the default NaN values when parsing the data.\n",
    "Depending on whether na_values is passed in, the behavior is as follows:\n",
    "\n",
    "- If keep_default_na is `True`, and na_values are specified, na_values\n",
    "  is appended to the default NaN values used for parsing.  \n",
    "- If keep_default_na is `True`, and na_values are not specified, only\n",
    "  the default NaN values are used for parsing.  \n",
    "- If keep_default_na is `False`, and na_values are specified, only\n",
    "  the NaN values specified na_values are used for parsing.  \n",
    "- If keep_default_na is `False`, and na_values are not specified, no\n",
    "  strings will be parsed as NaN.  \n",
    "\n",
    "\n",
    "Note that if na_filter is passed in as `False`, the keep_default_na and\n",
    "na_values parameters will be ignored.\n",
    "\n",
    "</dd>\n",
    "<dt>na_filter</dt>\n",
    "boolean, default `True`<dd>\n",
    "Detect missing value markers (empty strings and the value of na_values). In\n",
    "data without any NAs, passing `na_filter=False` can improve the performance\n",
    "of reading a large file.\n",
    "\n",
    "</dd>\n",
    "<dt>verbose</dt>\n",
    "boolean, default `False`<dd>\n",
    "Indicate number of NA values placed in non-numeric columns.\n",
    "\n",
    "</dd>\n",
    "<dt>skip_blank_lines</dt>\n",
    "boolean, default `True`<dd>\n",
    "If `True`, skip over blank lines rather than interpreting as NaN values.\n",
    "\n",
    "</dd>\n",
    "\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Datetime handling\n",
    "\n",
    "\n",
    "<dl style='margin: 20px 0;'>\n",
    "<dt>parse_dates</dt>\n",
    "boolean or list of ints or names or list of lists or dict, default `False`.<dd>\n",
    "- If `True` -> try parsing the index.  \n",
    "- If `[1, 2, 3]` ->  try parsing columns 1, 2, 3 each as a separate date\n",
    "  column.  \n",
    "- If `[[1, 3]]` -> combine columns 1 and 3 and parse as a single date\n",
    "  column.  \n",
    "- If `{'foo': [1, 3]}` -> parse columns 1, 3 as date and call result ‘foo’.\n",
    "  A fast-path exists for iso8601-formatted dates.  \n",
    "\n",
    "\n",
    "</dd>\n",
    "<dt>infer_datetime_format</dt>\n",
    "boolean, default `False`<dd>\n",
    "If `True` and parse_dates is enabled for a column, attempt to infer the\n",
    "datetime format to speed up the processing.\n",
    "\n",
    "</dd>\n",
    "<dt>keep_date_col</dt>\n",
    "boolean, default `False`<dd>\n",
    "If `True` and parse_dates specifies combining multiple columns then keep the\n",
    "original columns.\n",
    "\n",
    "</dd>\n",
    "<dt>date_parser</dt>\n",
    "function, default `None`<dd>\n",
    "Function to use for converting a sequence of string columns to an array of\n",
    "datetime instances. The default uses `dateutil.parser.parser` to do the\n",
    "conversion. pandas will try to call date_parser in three different ways,\n",
    "advancing to the next if an exception occurs: 1) Pass one or more arrays (as\n",
    "defined by parse_dates) as arguments; 2) concatenate (row-wise) the string\n",
    "values from the columns defined by parse_dates into a single array and pass\n",
    "that; and 3) call date_parser once for each row using one or more strings\n",
    "(corresponding to the columns defined by parse_dates) as arguments.\n",
    "\n",
    "</dd>\n",
    "<dt>dayfirst</dt>\n",
    "boolean, default `False`<dd>\n",
    "DD/MM format dates, international and European format.\n",
    "\n",
    "</dd>\n",
    "<dt>cache_dates</dt>\n",
    "boolean, default True<dd>\n",
    "If True, use a cache of unique, converted dates to apply the datetime\n",
    "conversion. May produce significant speed-up when parsing duplicate\n",
    "date strings, especially ones with timezone offsets.\n",
    "\n",
    "New in version 0.25.0.\n",
    "\n",
    "</dd>\n",
    "\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Iteration\n",
    "\n",
    "\n",
    "<dl style='margin: 20px 0;'>\n",
    "<dt>iterator</dt>\n",
    "boolean, default `False`<dd>\n",
    "Return TextFileReader object for iteration or getting chunks with\n",
    "`get_chunk()`.\n",
    "\n",
    "</dd>\n",
    "<dt>chunksize</dt>\n",
    "int, default `None`<dd>\n",
    "Return TextFileReader object for iteration. See [iterating and chunking](#io-chunking) below.\n",
    "\n",
    "</dd>\n",
    "\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Quoting, compression, and file format\n",
    "\n",
    "\n",
    "<dl style='margin: 20px 0;'>\n",
    "<dt>compression</dt>\n",
    "{`'infer'`, `'gzip'`, `'bz2'`, `'zip'`, `'xz'`, `None`}, default `'infer'`<dd>\n",
    "For on-the-fly decompression of on-disk data. If ‘infer’, then use gzip,\n",
    "bz2, zip, or xz if filepath_or_buffer is a string ending in ‘.gz’, ‘.bz2’,\n",
    "‘.zip’, or ‘.xz’, respectively, and no decompression otherwise. If using ‘zip’,\n",
    "the ZIP file must contain only one data file to be read in.\n",
    "Set to `None` for no decompression.\n",
    "\n",
    "New in version 0.18.1: support for ‘zip’ and ‘xz’ compression.\n",
    "\n",
    "Changed in version 0.24.0: ‘infer’ option added and set to default.\n",
    "\n",
    "</dd>\n",
    "<dt>thousands</dt>\n",
    "str, default `None`<dd>\n",
    "Thousands separator.\n",
    "\n",
    "</dd>\n",
    "<dt>decimal</dt>\n",
    "str, default `'.'`<dd>\n",
    "Character to recognize as decimal point. E.g. use `','` for European data.\n",
    "\n",
    "</dd>\n",
    "<dt>float_precision</dt>\n",
    "string, default None<dd>\n",
    "Specifies which converter the C engine should use for floating-point values.\n",
    "The options are `None` for the ordinary converter, `high` for the\n",
    "high-precision converter, and `round_trip` for the round-trip converter.\n",
    "\n",
    "</dd>\n",
    "<dt>lineterminator</dt>\n",
    "str (length 1), default `None`<dd>\n",
    "Character to break file into lines. Only valid with C parser.\n",
    "\n",
    "</dd>\n",
    "<dt>quotechar</dt>\n",
    "str (length 1)<dd>\n",
    "The character used to denote the start and end of a quoted item. Quoted items\n",
    "can include the delimiter and it will be ignored.\n",
    "\n",
    "</dd>\n",
    "<dt>quoting</dt>\n",
    "int or `csv.QUOTE_*` instance, default `0`<dd>\n",
    "Control field quoting behavior per `csv.QUOTE_*` constants. Use one of\n",
    "`QUOTE_MINIMAL` (0), `QUOTE_ALL` (1), `QUOTE_NONNUMERIC` (2) or\n",
    "`QUOTE_NONE` (3).\n",
    "\n",
    "</dd>\n",
    "<dt>doublequote</dt>\n",
    "boolean, default `True`<dd>\n",
    "When `quotechar` is specified and `quoting` is not `QUOTE_NONE`,\n",
    "indicate whether or not to interpret two consecutive `quotechar` elements\n",
    "**inside** a field as a single `quotechar` element.\n",
    "\n",
    "</dd>\n",
    "<dt>escapechar</dt>\n",
    "str (length 1), default `None`<dd>\n",
    "One-character string used to escape delimiter when quoting is `QUOTE_NONE`.\n",
    "\n",
    "</dd>\n",
    "<dt>comment</dt>\n",
    "str, default `None`<dd>\n",
    "Indicates remainder of line should not be parsed. If found at the beginning of\n",
    "a line, the line will be ignored altogether. This parameter must be a single\n",
    "character. Like empty lines (as long as `skip_blank_lines=True`), fully\n",
    "commented lines are ignored by the parameter header but not by skiprows.\n",
    "For example, if `comment='#'`, parsing ‘#empty\\na,b,c\\n1,2,3’ with\n",
    "header=0 will result in ‘a,b,c’ being treated as the header.\n",
    "\n",
    "</dd>\n",
    "<dt>encoding</dt>\n",
    "str, default `None`<dd>\n",
    "Encoding to use for UTF when reading/writing (e.g. `'utf-8'`). [List of\n",
    "Python standard encodings](https://docs.python.org/3/library/codecs.html#standard-encodings).\n",
    "\n",
    "</dd>\n",
    "<dt>dialect</dt>\n",
    "str or [`csv.Dialect`](https://docs.python.org/3/library/csv.html#csv.Dialect) instance, default `None`<dd>\n",
    "If provided, this parameter will override values (default or not) for the\n",
    "following parameters: delimiter, doublequote, escapechar,\n",
    "skipinitialspace, quotechar, and quoting. If it is necessary to\n",
    "override values, a ParserWarning will be issued. See [`csv.Dialect`](https://docs.python.org/3/library/csv.html#csv.Dialect)\n",
    "documentation for more details.\n",
    "\n",
    "</dd>\n",
    "\n",
    "</dl>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Error handling\n",
    "\n",
    "\n",
    "<dl style='margin: 20px 0;'>\n",
    "<dt>error_bad_lines</dt>\n",
    "boolean, default `True`<dd>\n",
    "Lines with too many fields (e.g. a csv line with too many commas) will by\n",
    "default cause an exception to be raised, and no `DataFrame` will be\n",
    "returned. If `False`, then these “bad lines” will dropped from the\n",
    "`DataFrame` that is returned. See [bad lines](#io-bad-lines)\n",
    "below.\n",
    "\n",
    "</dd>\n",
    "<dt>warn_bad_lines</dt>\n",
    "boolean, default `True`<dd>\n",
    "If error_bad_lines is `False`, and warn_bad_lines is `True`, a warning for\n",
    "each “bad line” will be output.\n",
    "\n",
    "</dd>\n",
    "\n",
    "</dl>\n",
    "\n",
    "\n",
    "<a id='io-dtypes'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Specifying column data types\n",
    "\n",
    "You can indicate the data type for the whole `DataFrame` or individual\n",
    "columns:\n",
    "\n",
    "Fortunately, pandas offers more than one way to ensure that your column(s)\n",
    "contain only one `dtype`. If you’re unfamiliar with these concepts, you can\n",
    "see [here](getting_started/basics.ipynb#basics-dtypes) to learn more about dtypes, and\n",
    "[here](getting_started/basics.ipynb#basics-object-conversion) to learn more about `object` conversion in\n",
    "pandas.\n",
    "\n",
    "For instance, you can use the `converters` argument\n",
    "of `read_csv()`:\n",
    "\n",
    "Or you can use the `to_numeric()` function to coerce the\n",
    "dtypes after reading in the data,\n",
    "\n",
    "which will convert all valid parsing to floats, leaving the invalid parsing\n",
    "as `NaN`.\n",
    "\n",
    "Ultimately, how you deal with reading in columns containing mixed dtypes\n",
    "depends on your specific needs. In the case above, if you wanted to `NaN` out\n",
    "the data anomalies, then `to_numeric()` is probably your best option.\n",
    "However, if you wanted for all the data to be coerced, no matter the type, then\n",
    "using the `converters` argument of `read_csv()` would certainly be\n",
    "worth trying.\n",
    "\n",
    "> New in version 0.20.0: support for the Python parser.\n",
    "\n",
    "The `dtype` option is supported by the ‘python’ engine.\n",
    "\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">In some cases, reading in abnormal data with columns containing mixed dtypes\n",
    "will result in an inconsistent dataset. If you rely on pandas to infer the\n",
    "dtypes of your columns, the parsing engine will go and infer the dtypes for\n",
    "different chunks of the data, rather than the whole dataset at once. Consequently,\n",
    "you can end up with column(s) with mixed dtypes. For example,\n",
    "\n",
    "will result with mixed_df containing an `int` dtype for certain chunks\n",
    "of the column, and `str` for others due to the mixed dtypes from the\n",
    "data that was read in. It is important to note that the overall column will be\n",
    "marked with a `dtype` of `object`, which is used for columns with mixed dtypes.\n",
    "\n",
    "\n",
    "<a id='io-categorical'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Specifying categorical dtype\n",
    "\n",
    "New in version 0.19.0.\n",
    "\n",
    "`Categorical` columns can be parsed directly by specifying `dtype='category'` or\n",
    "`dtype=CategoricalDtype(categories, ordered)`.\n",
    "\n",
    "Individual columns can be parsed as a `Categorical` using a dict\n",
    "specification:\n",
    "\n",
    "New in version 0.21.0.\n",
    "\n",
    "Specifying `dtype='category'` will result in an unordered `Categorical`\n",
    "whose `categories` are the unique values observed in the data. For more\n",
    "control on the categories and order, create a\n",
    "`CategoricalDtype` ahead of time, and pass that for\n",
    "that column’s `dtype`.\n",
    "\n",
    "When using `dtype=CategoricalDtype`, “unexpected” values outside of\n",
    "`dtype.categories` are treated as missing values.\n",
    "\n",
    "This matches the behavior of `Categorical.set_categories()`.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">With `dtype='category'`, the resulting categories will always be parsed\n",
    "as strings (object dtype). If the categories are numeric they can be\n",
    "converted using the `to_numeric()` function, or as appropriate, another\n",
    "converter such as `to_datetime()`.\n",
    "\n",
    "When `dtype` is a `CategoricalDtype` with homogeneous `categories` (\n",
    "all numeric, all datetimes, etc.), the conversion is done automatically."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Naming and using columns\n",
    "\n",
    "\n",
    "<a id='io-headers'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Handling column names\n",
    "\n",
    "A file may or may not have a header row. pandas assumes the first row should be\n",
    "used as the column names:\n",
    "\n",
    "By specifying the `names` argument in conjunction with `header` you can\n",
    "indicate other names to use and whether or not to throw away the header row (if\n",
    "any):\n",
    "\n",
    "If the header is in a row other than the first, pass the row number to\n",
    "`header`. This will skip the preceding rows:\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">Default behavior is to infer the column names: if no names are\n",
    "passed the behavior is identical to `header=0` and column names\n",
    "are inferred from the first non-blank line of the file, if column\n",
    "names are passed explicitly then the behavior is identical to\n",
    "`header=None`.\n",
    "\n",
    "\n",
    "<a id='io-dupe-names'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Duplicate names parsing\n",
    "\n",
    "If the file or header contains duplicate names, pandas will by default\n",
    "distinguish between them so as to prevent overwriting data:\n",
    "\n",
    "There is no more duplicate data because `mangle_dupe_cols=True` by default,\n",
    "which modifies a series of duplicate columns ‘X’, …, ‘X’ to become\n",
    "‘X’, ‘X.1’, …, ‘X.N’.  If `mangle_dupe_cols=False`, duplicate data can\n",
    "arise:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hide-output": false
   },
   "source": [
    "```ipython\n",
    "In [2]: data = 'a,b,a\\n0,1,2\\n3,4,5'\n",
    "In [3]: pd.read_csv(StringIO(data), mangle_dupe_cols=False)\n",
    "Out[3]:\n",
    "   a  b  a\n",
    "0  2  1  2\n",
    "1  5  4  5\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To prevent users from encountering this problem with duplicate data, a `ValueError`\n",
    "exception is raised if `mangle_dupe_cols != True`:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hide-output": false
   },
   "source": [
    "```ipython\n",
    "In [2]: data = 'a,b,a\\n0,1,2\\n3,4,5'\n",
    "In [3]: pd.read_csv(StringIO(data), mangle_dupe_cols=False)\n",
    "...\n",
    "ValueError: Setting mangle_dupe_cols=False is not supported yet\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<a id='io-usecols'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Filtering columns (`usecols`)\n",
    "\n",
    "The `usecols` argument allows you to select any subset of the columns in a\n",
    "file, either using the column names, position numbers or a callable:\n",
    "\n",
    "New in version 0.20.0: support for callable usecols arguments\n",
    "\n",
    "The `usecols` argument can also be used to specify which columns not to\n",
    "use in the final result:\n",
    "\n",
    "In this case, the callable is specifying that we exclude the “a” and “c”\n",
    "columns from the output."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comments and empty lines\n",
    "\n",
    "\n",
    "<a id='io-skiplines'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Ignoring line comments and empty lines\n",
    "\n",
    "If the `comment` parameter is specified, then completely commented lines will\n",
    "be ignored. By default, completely blank lines will be ignored as well.\n",
    "\n",
    "If `skip_blank_lines=False`, then `read_csv` will not ignore blank lines:\n",
    "\n",
    "The presence of ignored lines might create ambiguities involving line numbers;\n",
    "the parameter `header` uses row numbers (ignoring commented/empty\n",
    "lines), while `skiprows` uses line numbers (including commented/empty lines):\n",
    "\n",
    "If both `header` and `skiprows` are specified, `header` will be\n",
    "relative to the end of `skiprows`. For example:\n",
    "\n",
    "\n",
    "<a id='io-comments'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Comments\n",
    "\n",
    "Sometimes comments or meta data may be included in a file:\n",
    "\n",
    "By default, the parser includes the comments in the output:\n",
    "\n",
    "We can suppress the comments using the `comment` keyword:\n",
    "\n",
    "\n",
    "<a id='io-unicode'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dealing with Unicode data\n",
    "\n",
    "The `encoding` argument should be used for encoded unicode data, which will\n",
    "result in byte strings being decoded to unicode in the result:\n",
    "\n",
    "Some formats which encode all characters as multiple bytes, like UTF-16, won’t\n",
    "parse correctly at all without specifying the encoding. [Full list of Python\n",
    "standard encodings](https://docs.python.org/3/library/codecs.html#standard-encodings).\n",
    "\n",
    "\n",
    "<a id='io-index-col'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Index columns and trailing delimiters\n",
    "\n",
    "If a file has one more column of data than the number of column names, the\n",
    "first column will be used as the `DataFrame`’s row names:\n",
    "\n",
    "Ordinarily, you can achieve this behavior using the `index_col` option.\n",
    "\n",
    "There are some exception cases when a file has been prepared with delimiters at\n",
    "the end of each data line, confusing the parser. To explicitly disable the\n",
    "index column inference and discard the last column, pass `index_col=False`:\n",
    "\n",
    "If a subset of data is being parsed using the `usecols` option, the\n",
    "`index_col` specification is based on that subset, not the original data.\n",
    "\n",
    "\n",
    "<a id='io-parse-dates'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Date Handling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Specifying date columns\n",
    "\n",
    "To better facilitate working with datetime data, `read_csv()`\n",
    "uses the keyword arguments `parse_dates` and `date_parser`\n",
    "to allow users to specify a variety of columns and date/time formats to turn the\n",
    "input text data into `datetime` objects.\n",
    "\n",
    "The simplest case is to just pass in `parse_dates=True`:\n",
    "\n",
    "It is often the case that we may want to store date and time data separately,\n",
    "or store various date fields separately. the `parse_dates` keyword can be\n",
    "used to specify a combination of columns to parse the dates and/or times from.\n",
    "\n",
    "You can specify a list of column lists to `parse_dates`, the resulting date\n",
    "columns will be prepended to the output (so as to not affect the existing column\n",
    "order) and the new column names will be the concatenation of the component\n",
    "column names:\n",
    "\n",
    "By default the parser removes the component date columns, but you can choose\n",
    "to retain them via the `keep_date_col` keyword:\n",
    "\n",
    "Note that if you wish to combine multiple columns into a single date column, a\n",
    "nested list must be used. In other words, `parse_dates=[1, 2]` indicates that\n",
    "the second and third columns should each be parsed as separate date columns\n",
    "while `parse_dates=[[1, 2]]` means the two columns should be parsed into a\n",
    "single column.\n",
    "\n",
    "You can also use a dict to specify custom name columns:\n",
    "\n",
    "It is important to remember that if multiple text columns are to be parsed into\n",
    "a single date column, then a new column is prepended to the data. The index_col\n",
    "specification is based off of this new set of columns rather than the original\n",
    "data columns:\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">If a column or index contains an unparsable date, the entire column or\n",
    "index will be returned unaltered as an object data type. For non-standard\n",
    "datetime parsing, use `to_datetime()` after `pd.read_csv`.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">read_csv has a fast_path for parsing datetime strings in iso8601 format,\n",
    "e.g “2000-01-01T00:01:02+00:00” and similar variations. If you can arrange\n",
    "for your data to store datetimes in this format, load times will be\n",
    "significantly faster, ~20x has been observed.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">When passing a dict as the parse_dates argument, the order of\n",
    "the columns prepended is not guaranteed, because dict objects do not impose\n",
    "an ordering on their keys. On Python 2.7+ you may use collections.OrderedDict\n",
    "instead of a regular dict if this matters to you. Because of this, when using a\n",
    "dict for ‘parse_dates’ in conjunction with the index_col argument, it’s best to\n",
    "specify index_col as a column label rather then as an index on the resulting frame."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Date parsing functions\n",
    "\n",
    "Finally, the parser allows you to specify a custom `date_parser` function to\n",
    "take full advantage of the flexibility of the date parsing API:\n",
    "\n",
    "Pandas will try to call the `date_parser` function in three different ways. If\n",
    "an exception is raised, the next one is tried:\n",
    "\n",
    "1. `date_parser` is first called with one or more arrays as arguments,\n",
    "  as defined using parse_dates (e.g., `date_parser(['2013', '2013'], ['1', '2'])`).  \n",
    "1. If #1 fails, `date_parser` is called with all the columns\n",
    "  concatenated row-wise into a single array (e.g., `date_parser(['2013 1', '2013 2'])`).  \n",
    "1. If #2 fails, `date_parser` is called once for every row with one or more\n",
    "  string arguments from the columns indicated with parse_dates\n",
    "  (e.g., `date_parser('2013', '1')` for the first row, `date_parser('2013', '2')`\n",
    "  for the second, etc.).  \n",
    "\n",
    "\n",
    "Note that performance-wise, you should try these methods of parsing dates in order:\n",
    "\n",
    "1. Try to infer the format using `infer_datetime_format=True` (see section below).  \n",
    "1. If you know the format, use `pd.to_datetime()`:\n",
    "  `date_parser=lambda x: pd.to_datetime(x, format=...)`.  \n",
    "1. If you have a really non-standard format, use a custom `date_parser` function.\n",
    "  For optimal performance, this should be vectorized, i.e., it should accept arrays\n",
    "  as arguments.  \n",
    "\n",
    "\n",
    "You can explore the date parsing functionality in\n",
    "[date_converters.py](https://github.com/pandas-dev/pandas/blob/master/pandas/io/date_converters.py)\n",
    "and add your own. We would love to turn this module into a community supported\n",
    "set of date/time parsers. To get you started, `date_converters.py` contains\n",
    "functions to parse dual date and time columns, year/month/day columns,\n",
    "and year/month/day/hour/minute/second columns. It also contains a\n",
    "`generic_parser` function so you can curry it with a function that deals with\n",
    "a single date rather than the entire array.\n",
    "\n",
    "\n",
    "<a id='io-csv-mixed-timezones'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Parsing a CSV with mixed timezones\n",
    "\n",
    "Pandas cannot natively represent a column or index with mixed timezones. If your CSV\n",
    "file contains columns with a mixture of timezones, the default result will be\n",
    "an object-dtype column with strings, even with `parse_dates`.\n",
    "\n",
    "To parse the mixed-timezone values as a datetime column, pass a partially-applied\n",
    "`to_datetime()` with `utc=True` as the `date_parser`.\n",
    "\n",
    "\n",
    "<a id='io-dayfirst'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Inferring datetime format\n",
    "\n",
    "If you have `parse_dates` enabled for some or all of your columns, and your\n",
    "datetime strings are all formatted the same way, you may get a large speed\n",
    "up by setting `infer_datetime_format=True`.  If set, pandas will attempt\n",
    "to guess the format of your datetime strings, and then use a faster means\n",
    "of parsing the strings.  5-10x parsing speeds have been observed.  pandas\n",
    "will fallback to the usual parsing if either the format cannot be guessed\n",
    "or the format that was guessed cannot properly parse the entire column\n",
    "of strings.  So in general, `infer_datetime_format` should not have any\n",
    "negative consequences if enabled.\n",
    "\n",
    "Here are some examples of datetime strings that can be guessed (All\n",
    "representing December 30th, 2011 at 00:00:00):\n",
    "\n",
    "- “20111230”  \n",
    "- “2011/12/30”  \n",
    "- “20111230 00:00:00”  \n",
    "- “12/30/2011 00:00:00”  \n",
    "- “30/Dec/2011 00:00:00”  \n",
    "- “30/December/2011 00:00:00”  \n",
    "\n",
    "\n",
    "Note that `infer_datetime_format` is sensitive to `dayfirst`.  With\n",
    "`dayfirst=True`, it will guess “01/12/2011” to be December 1st. With\n",
    "`dayfirst=False` (default) it will guess “01/12/2011” to be January 12th."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### International date formats\n",
    "\n",
    "While US date formats tend to be MM/DD/YYYY, many international formats use\n",
    "DD/MM/YYYY instead. For convenience, a `dayfirst` keyword is provided:\n",
    "\n",
    "\n",
    "<a id='io-float-precision'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Specifying method for floating-point conversion\n",
    "\n",
    "The parameter `float_precision` can be specified in order to use\n",
    "a specific floating-point converter during parsing with the C engine.\n",
    "The options are the ordinary converter, the high-precision converter, and\n",
    "the round-trip converter (which is guaranteed to round-trip values after\n",
    "writing to a file). For example:\n",
    "\n",
    "\n",
    "<a id='io-thousands'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Thousand separators\n",
    "\n",
    "For large numbers that have been written with a thousands separator, you can\n",
    "set the `thousands` keyword to a string of length 1 so that integers will be parsed\n",
    "correctly:\n",
    "\n",
    "By default, numbers with a thousands separator will be parsed as strings:\n",
    "\n",
    "The `thousands` keyword allows integers to be parsed correctly:\n",
    "\n",
    "\n",
    "<a id='io-na-values'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### NA values\n",
    "\n",
    "To control which values are parsed as missing values (which are signified by\n",
    "`NaN`), specify a string in `na_values`. If you specify a list of strings,\n",
    "then all values in it are considered to be missing values. If you specify a\n",
    "number (a `float`, like `5.0` or an `integer` like `5`), the\n",
    "corresponding equivalent values will also imply a missing value (in this case\n",
    "effectively `[5.0, 5]` are recognized as `NaN`).\n",
    "\n",
    "To completely override the default values that are recognized as missing, specify `keep_default_na=False`.\n",
    "\n",
    "\n",
    "<a id='io-navaluesconst'></a>\n",
    "The default `NaN` recognized values are `['-1.#IND', '1.#QNAN', '1.#IND', '-1.#QNAN', '#N/A N/A', '#N/A', 'N/A',\n",
    "'n/a', 'NA', '#NA', 'NULL', 'null', 'NaN', '-NaN', 'nan', '-nan', '']`.\n",
    "\n",
    "Let us consider some examples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "pd.read_csv('path_to_file.csv', na_values=[5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the example above `5` and `5.0` will be recognized as `NaN`, in\n",
    "addition to the defaults. A string will first be interpreted as a numerical\n",
    "`5`, then as a `NaN`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "pd.read_csv('path_to_file.csv', keep_default_na=False, na_values=[\"\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above, only an empty field will be recognized as `NaN`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "pd.read_csv('path_to_file.csv', keep_default_na=False, na_values=[\"NA\", \"0\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above, both `NA` and `0` as strings are `NaN`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "pd.read_csv('path_to_file.csv', na_values=[\"Nope\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The default values, in addition to the string `\"Nope\"` are recognized as\n",
    "`NaN`.\n",
    "\n",
    "\n",
    "<a id='io-infinity'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Infinity\n",
    "\n",
    "`inf` like values will be parsed as `np.inf` (positive infinity), and `-inf` as `-np.inf` (negative infinity).\n",
    "These will ignore the case of the value, meaning `Inf`, will also be parsed as `np.inf`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Returning Series\n",
    "\n",
    "Using the `squeeze` keyword, the parser will return output with a single column\n",
    "as a `Series`:\n",
    "\n",
    "\n",
    "<a id='io-boolean'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Boolean values\n",
    "\n",
    "The common values `True`, `False`, `TRUE`, and `FALSE` are all\n",
    "recognized as boolean. Occasionally you might want to recognize other values\n",
    "as being boolean. To do this, use the `true_values` and `false_values`\n",
    "options as follows:\n",
    "\n",
    "\n",
    "<a id='io-bad-lines'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Handling “bad” lines\n",
    "\n",
    "Some files may have malformed lines with too few fields or too many. Lines with\n",
    "too few fields will have NA values filled in the trailing fields. Lines with\n",
    "too many fields will raise an error by default:\n",
    "\n",
    "You can elect to skip bad lines:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hide-output": false
   },
   "source": [
    "```ipython\n",
    "In [29]: pd.read_csv(StringIO(data), error_bad_lines=False)\n",
    "Skipping line 3: expected 3 fields, saw 4\n",
    "\n",
    "Out[29]:\n",
    "   a  b   c\n",
    "0  1  2   3\n",
    "1  8  9  10\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can also use the `usecols` parameter to eliminate extraneous column\n",
    "data that appear in some lines but not others:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hide-output": false
   },
   "source": [
    "```ipython\n",
    "In [30]: pd.read_csv(StringIO(data), usecols=[0, 1, 2])\n",
    "\n",
    " Out[30]:\n",
    "    a  b   c\n",
    " 0  1  2   3\n",
    " 1  4  5   6\n",
    " 2  8  9  10\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<a id='io-dialect'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dialect\n",
    "\n",
    "The `dialect` keyword gives greater flexibility in specifying the file format.\n",
    "By default it uses the Excel dialect but you can specify either the dialect name\n",
    "or a [`csv.Dialect`](https://docs.python.org/3/library/csv.html#csv.Dialect) instance.\n",
    "\n",
    "Suppose you had data with unenclosed quotes:\n",
    "\n",
    "By default, `read_csv` uses the Excel dialect and treats the double quote as\n",
    "the quote character, which causes it to fail when it finds a newline before it\n",
    "finds the closing double quote.\n",
    "\n",
    "We can get around this using `dialect`:\n",
    "\n",
    "All of the dialect options can be specified separately by keyword arguments:\n",
    "\n",
    "Another common dialect option is `skipinitialspace`, to skip any whitespace\n",
    "after a delimiter:\n",
    "\n",
    "The parsers make every attempt to “do the right thing” and not be fragile. Type\n",
    "inference is a pretty big deal. If a column can be coerced to integer dtype\n",
    "without altering the contents, the parser will do so. Any non-numeric\n",
    "columns will come through as object dtype as with the rest of pandas objects.\n",
    "\n",
    "\n",
    "<a id='io-quoting'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quoting and Escape Characters\n",
    "\n",
    "Quotes (and other escape characters) in embedded fields can be handled in any\n",
    "number of ways. One way is to use backslashes; to properly parse this data, you\n",
    "should pass the `escapechar` option:\n",
    "\n",
    "\n",
    "<a id='io-fwf'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Files with fixed width columns\n",
    "\n",
    "While `read_csv()` reads delimited data, the `read_fwf()` function works\n",
    "with data files that have known and fixed column widths. The function parameters\n",
    "to `read_fwf` are largely the same as read_csv with two extra parameters, and\n",
    "a different usage of the `delimiter` parameter:\n",
    "\n",
    "- `colspecs`: A list of pairs (tuples) giving the extents of the\n",
    "  fixed-width fields of each line as half-open intervals (i.e.,  [from, to[ ).\n",
    "  String value ‘infer’ can be used to instruct the parser to try detecting\n",
    "  the column specifications from the first 100 rows of the data. Default\n",
    "  behavior, if not specified, is to infer.  \n",
    "- `widths`: A list of field widths which can be used instead of ‘colspecs’\n",
    "  if the intervals are contiguous.  \n",
    "- `delimiter`: Characters to consider as filler characters in the fixed-width file.\n",
    "  Can be used to specify the filler character of the fields\n",
    "  if it is not spaces (e.g., ‘~’).  \n",
    "\n",
    "\n",
    "Consider a typical fixed-width data file:\n",
    "\n",
    "In order to parse this file into a `DataFrame`, we simply need to supply the\n",
    "column specifications to the read_fwf function along with the file name:\n",
    "\n",
    "Note how the parser automatically picks column names X.<column number> when\n",
    "`header=None` argument is specified. Alternatively, you can supply just the\n",
    "column widths for contiguous columns:\n",
    "\n",
    "The parser will take care of extra white spaces around the columns\n",
    "so it’s ok to have extra separation between the columns in the file.\n",
    "\n",
    "By default, `read_fwf` will try to infer the file’s `colspecs` by using the\n",
    "first 100 rows of the file. It can do it only in cases when the columns are\n",
    "aligned and correctly separated by the provided `delimiter` (default delimiter\n",
    "is whitespace).\n",
    "\n",
    "New in version 0.20.0.\n",
    "\n",
    "`read_fwf` supports the `dtype` parameter for specifying the types of\n",
    "parsed columns to be different from the inferred type."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Indexes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Files with an “implicit” index column\n",
    "\n",
    "Consider a file with one less entry in the header than the number of data\n",
    "column:\n",
    "\n",
    "In this special case, `read_csv` assumes that the first column is to be used\n",
    "as the index of the `DataFrame`:\n",
    "\n",
    "Note that the dates weren’t automatically parsed. In that case you would need\n",
    "to do as before:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reading an index with a `MultiIndex`\n",
    "\n",
    "\n",
    "<a id='io-csv-multiindex'></a>\n",
    "Suppose you have data indexed by two columns:\n",
    "\n",
    "The `index_col` argument to `read_csv` can take a list of\n",
    "column numbers to turn multiple columns into a `MultiIndex` for the index of the\n",
    "returned object:\n",
    "\n",
    "\n",
    "<a id='io-multi-index-columns'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reading columns with a `MultiIndex`\n",
    "\n",
    "By specifying list of row locations for the `header` argument, you\n",
    "can read in a `MultiIndex` for the columns. Specifying non-consecutive\n",
    "rows will skip the intervening rows.\n",
    "\n",
    "`read_csv` is also able to interpret a more common format\n",
    "of multi-columns indices.\n",
    "\n",
    "Note: If an `index_col` is not specified (e.g. you don’t have an index, or wrote it\n",
    "with `df.to_csv(..., index=False)`, then any `names` on the columns index will be *lost*.\n",
    "\n",
    "\n",
    "<a id='io-sniff'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Automatically “sniffing” the delimiter\n",
    "\n",
    "`read_csv` is capable of inferring delimited (not necessarily\n",
    "comma-separated) files, as pandas uses the [`csv.Sniffer`](https://docs.python.org/3/library/csv.html#csv.Sniffer)\n",
    "class of the csv module. For this, you have to specify `sep=None`.\n",
    "\n",
    "\n",
    "<a id='io-multiple-files'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading multiple files to create a single DataFrame\n",
    "\n",
    "It’s best to use `concat()` to combine multiple files.\n",
    "See the [cookbook](cookbook.ipynb#cookbook-csv-multiple-files) for an example.\n",
    "\n",
    "\n",
    "<a id='io-chunking'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Iterating through files chunk by chunk\n",
    "\n",
    "Suppose you wish to iterate through a (potentially very large) file lazily\n",
    "rather than reading the entire file into memory, such as the following:\n",
    "\n",
    "By specifying a `chunksize` to `read_csv`, the return\n",
    "value will be an iterable object of type `TextFileReader`:\n",
    "\n",
    "Specifying `iterator=True` will also return the `TextFileReader` object:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Specifying the parser engine\n",
    "\n",
    "Under the hood pandas uses a fast and efficient parser implemented in C as well\n",
    "as a Python implementation which is currently more feature-complete. Where\n",
    "possible pandas uses the C parser (specified as `engine='c'`), but may fall\n",
    "back to Python if C-unsupported options are specified. Currently, C-unsupported\n",
    "options include:\n",
    "\n",
    "- `sep` other than a single character (e.g. regex separators)  \n",
    "- `skipfooter`  \n",
    "- `sep=None` with `delim_whitespace=False`  \n",
    "\n",
    "\n",
    "Specifying any of the above options will produce a `ParserWarning` unless the\n",
    "python engine is selected explicitly using `engine='python'`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading remote files\n",
    "\n",
    "You can pass in a URL to a CSV file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "df = pd.read_csv('https://download.bls.gov/pub/time.series/cu/cu.item',\n",
    "                 sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "S3 URLs are handled as well but require installing the [S3Fs](https://pypi.org/project/s3fs/) library:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "df = pd.read_csv('s3://pandas-test/tips.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If your S3 bucket requires credentials you will need to set them as environment\n",
    "variables or in the `~/.aws/credentials` config file, refer to the [S3Fs\n",
    "documentation on credentials](https://s3fs.readthedocs.io/en/latest/#credentials)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing out data\n",
    "\n",
    "\n",
    "<a id='io-store-in-csv'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Writing to CSV format\n",
    "\n",
    "The `Series` and `DataFrame` objects have an instance method `to_csv` which\n",
    "allows storing the contents of the object as a comma-separated-values file. The\n",
    "function takes a number of arguments. Only the first is required.\n",
    "\n",
    "- `path_or_buf`: A string path to the file to write or a file object.  If a file object it must be opened with newline=’‘  \n",
    "- `sep` : Field delimiter for the output file (default “,”)  \n",
    "- `na_rep`: A string representation of a missing value (default ‘’)  \n",
    "- `float_format`: Format string for floating point numbers  \n",
    "- `columns`: Columns to write (default None)  \n",
    "- `header`: Whether to write out the column names (default True)  \n",
    "- `index`: whether to write row (index) names (default True)  \n",
    "- `index_label`: Column label(s) for index column(s) if desired. If None\n",
    "  (default), and header and index are True, then the index names are\n",
    "  used. (A sequence should be given if the `DataFrame` uses MultiIndex).  \n",
    "- `mode` : Python write mode, default ‘w’  \n",
    "- `encoding`: a string representing the encoding to use if the contents are\n",
    "  non-ASCII, for Python versions prior to 3  \n",
    "- `line_terminator`: Character sequence denoting line end (default os.linesep)  \n",
    "- `quoting`: Set quoting rules as in csv module (default csv.QUOTE_MINIMAL). Note that if you have set a float_format then floats are converted to strings and csv.QUOTE_NONNUMERIC will treat them as non-numeric  \n",
    "- `quotechar`: Character used to quote fields (default ‘”’)  \n",
    "- `doublequote`: Control quoting of `quotechar` in fields (default True)  \n",
    "- `escapechar`: Character used to escape `sep` and `quotechar` when\n",
    "  appropriate (default None)  \n",
    "- `chunksize`: Number of rows to write at a time  \n",
    "- `date_format`: Format string for datetime objects  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Writing a formatted string\n",
    "\n",
    "\n",
    "<a id='io-formatting'></a>\n",
    "The `DataFrame` object has an instance method `to_string` which allows control\n",
    "over the string representation of the object. All arguments are optional:\n",
    "\n",
    "- `buf` default None, for example a StringIO object  \n",
    "- `columns` default None, which columns to write  \n",
    "- `col_space` default None, minimum width of each column.  \n",
    "- `na_rep` default `NaN`, representation of NA value  \n",
    "- `formatters` default None, a dictionary (by column) of functions each of\n",
    "  which takes a single argument and returns a formatted string  \n",
    "- `float_format` default None, a function which takes a single (float)\n",
    "  argument and returns a formatted string; to be applied to floats in the\n",
    "  `DataFrame`.  \n",
    "- `sparsify` default True, set to False for a `DataFrame` with a hierarchical\n",
    "  index to print every MultiIndex key at each row.  \n",
    "- `index_names` default True, will print the names of the indices  \n",
    "- `index` default True, will print the index (ie, row labels)  \n",
    "- `header` default True, will print the column labels  \n",
    "- `justify` default `left`, will print column headers left- or\n",
    "  right-justified  \n",
    "\n",
    "\n",
    "The `Series` object also has a `to_string` method, but with only the `buf`,\n",
    "`na_rep`, `float_format` arguments. There is also a `length` argument\n",
    "which, if set to `True`, will additionally output the length of the Series.\n",
    "\n",
    "\n",
    "<a id='io-json'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## JSON\n",
    "\n",
    "Read and write `JSON` format files and strings.\n",
    "\n",
    "\n",
    "<a id='io-json-writer'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing JSON\n",
    "\n",
    "A `Series` or `DataFrame` can be converted to a valid JSON string. Use `to_json`\n",
    "with optional parameters:\n",
    "\n",
    "- `path_or_buf` : the pathname or buffer to write the output\n",
    "  This can be `None` in which case a JSON string is returned  \n",
    "- `orient` :  \n",
    "  \n",
    "  <dl style='margin: 20px 0;'>\n",
    "  <dt>`Series`:</dt>\n",
    "  <dd>\n",
    "  - default is `index`  \n",
    "  - allowed values are {`split`, `records`, `index`}  \n",
    "  </dd>\n",
    "  <dt>`DataFrame`:</dt>\n",
    "  <dd>\n",
    "  - default is `columns`  \n",
    "  - allowed values are {`split`, `records`, `index`, `columns`, `values`, `table`}  \n",
    "  </dd>\n",
    "  \n",
    "  </dl>\n",
    "  \n",
    "  The format of the JSON string  \n",
    "  ``  \n",
    "    \n",
    "  ``  \n",
    "    \n",
    "  ``  \n",
    "    \n",
    "  ``  \n",
    "    \n",
    "  ``  \n",
    "    \n",
    "  |split|dict like {index -> [index], columns -> [columns], data -> [values]}|\n",
    "  |records|list like [{column -> value}, … , {column -> value}]|\n",
    "  |index|dict like {index -> {column -> value}}|\n",
    "  |columns|dict like {column -> {index -> value}}|\n",
    "  |values|just the values array|\n",
    "- `date_format` : string, type of date conversion, ‘epoch’ for timestamp, ‘iso’ for ISO8601.  \n",
    "- `double_precision` : The number of decimal places to use when encoding floating point values, default 10.  \n",
    "- `force_ascii` : force encoded string to be ASCII, default True.  \n",
    "- `date_unit` : The time unit to encode to, governs timestamp and ISO8601 precision. One of ‘s’, ‘ms’, ‘us’ or ‘ns’ for seconds, milliseconds, microseconds and nanoseconds respectively. Default ‘ms’.  \n",
    "- `default_handler` : The handler to call if an object cannot otherwise be converted to a suitable format for JSON. Takes a single argument, which is the object to convert, and returns a serializable object.  \n",
    "- `lines` : If `records` orient, then will write each record per line as json.  \n",
    "\n",
    "\n",
    "Note `NaN`’s, `NaT`’s and `None` will be converted to `null` and `datetime` objects will be converted based on the `date_format` and `date_unit` parameters."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Orient options\n",
    "\n",
    "There are a number of different options for the format of the resulting JSON\n",
    "file / string. Consider the following `DataFrame` and `Series`:\n",
    "\n",
    "**Column oriented** (the default for `DataFrame`) serializes the data as\n",
    "nested JSON objects with column labels acting as the primary index:\n",
    "\n",
    "**Index oriented** (the default for `Series`) similar to column oriented\n",
    "but the index labels are now primary:\n",
    "\n",
    "**Record oriented** serializes the data to a JSON array of column -> value records,\n",
    "index labels are not included. This is useful for passing `DataFrame` data to plotting\n",
    "libraries, for example the JavaScript library `d3.js`:\n",
    "\n",
    "**Value oriented** is a bare-bones option which serializes to nested JSON arrays of\n",
    "values only, column and index labels are not included:\n",
    "\n",
    "**Split oriented** serializes to a JSON object containing separate entries for\n",
    "values, index and columns. Name is also included for `Series`:\n",
    "\n",
    "**Table oriented** serializes to the JSON [Table Schema](https://specs.frictionlessdata.io/json-table-schema/), allowing for the\n",
    "preservation of metadata including but not limited to dtypes and index names.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">Any orient option that encodes to a JSON object will not preserve the ordering of\n",
    "index and column labels during round-trip serialization. If you wish to preserve\n",
    "label ordering use the split option as it uses ordered containers."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Date handling\n",
    "\n",
    "Writing in ISO date format:\n",
    "\n",
    "Writing in ISO date format, with microseconds:\n",
    "\n",
    "Epoch timestamps, in seconds:\n",
    "\n",
    "Writing to a file, with a date index and a date column:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fallback behavior\n",
    "\n",
    "If the JSON serializer cannot handle the container contents directly it will\n",
    "fall back in the following manner:\n",
    "\n",
    "- if the dtype is unsupported (e.g. `np.complex`) then the `default_handler`, if provided, will be called\n",
    "  for each value, otherwise an exception is raised.  \n",
    "- if an object is unsupported it will attempt the following:  \n",
    "  \n",
    "  - check if the object has defined a `toDict` method and call it.\n",
    "    A `toDict` method should return a `dict` which will then be JSON serialized.  \n",
    "  - invoke the `default_handler` if one was provided.  \n",
    "  - convert the object to a `dict` by traversing its contents. However this will often fail\n",
    "    with an `OverflowError` or give unexpected results.  \n",
    "  \n",
    "\n",
    "\n",
    "In general the best approach for unsupported objects or dtypes is to provide a `default_handler`.\n",
    "For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    ">>> DataFrame([1.0, 2.0, complex(1.0, 2.0)]).to_json()  # raises\n",
    "RuntimeError: Unhandled numpy dtype 15"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "can be dealt with by specifying a simple `default_handler`:\n",
    "\n",
    "\n",
    "<a id='io-json-reader'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading JSON\n",
    "\n",
    "Reading a JSON string to pandas object can take a number of parameters.\n",
    "The parser will try to parse a `DataFrame` if `typ` is not supplied or\n",
    "is `None`. To explicitly force `Series` parsing, pass `typ=series`\n",
    "\n",
    "- `filepath_or_buffer` : a **VALID** JSON string or file handle / StringIO. The string could be\n",
    "  a URL. Valid URL schemes include http, ftp, S3, and file. For file URLs, a host\n",
    "  is expected. For instance, a local file could be\n",
    "  file ://localhost/path/to/table.json  \n",
    "- `typ`    : type of object to recover (series or frame), default ‘frame’  \n",
    "- `orient` :  \n",
    "  \n",
    "  <dl style='margin: 20px 0;'>\n",
    "  <dt>Series :</dt>\n",
    "  <dd>\n",
    "  - default is `index`  \n",
    "  - allowed values are {`split`, `records`, `index`}  \n",
    "  </dd>\n",
    "  <dt>DataFrame</dt>\n",
    "  <dd>\n",
    "  - default is `columns`  \n",
    "  - allowed values are {`split`, `records`, `index`, `columns`, `values`, `table`}  \n",
    "  </dd>\n",
    "  \n",
    "  </dl>\n",
    "  \n",
    "  The format of the JSON string  \n",
    "  ``  \n",
    "    \n",
    "  ``  \n",
    "    \n",
    "  ``  \n",
    "    \n",
    "  ``  \n",
    "    \n",
    "  ``  \n",
    "    \n",
    "  ``  \n",
    "  [](https://specs.frictionlessdata.io/json-table-schema/)  \n",
    "  |split|dict like {index -> [index], columns -> [columns], data -> [values]}|\n",
    "  |records|list like [{column -> value}, … , {column -> value}]|\n",
    "  |index|dict like {index -> {column -> value}}|\n",
    "  |columns|dict like {column -> {index -> value}}|\n",
    "  |values|just the values array|\n",
    "  |table|adhering to the JSON Table Schema|\n",
    "- `dtype` : if True, infer dtypes, if a dict of column to dtype, then use those, if `False`, then don’t infer dtypes at all, default is True, apply only to the data.  \n",
    "- `convert_axes` : boolean, try to convert the axes to the proper dtypes, default is `True`  \n",
    "- `convert_dates` : a list of columns to parse for dates; If `True`, then try to parse date-like columns, default is `True`.  \n",
    "- `keep_default_dates` : boolean, default `True`. If parsing dates, then parse the default date-like columns.  \n",
    "- `numpy` : direct decoding to NumPy arrays. default is `False`;\n",
    "  Supports numeric data only, although labels may be non-numeric. Also note that the JSON ordering **MUST** be the same for each term if `numpy=True`.  \n",
    "- `precise_float` : boolean, default `False`. Set to enable usage of higher precision (strtod) function when decoding string to double values. Default (`False`) is to use fast but less precise builtin functionality.  \n",
    "- `date_unit` : string, the timestamp unit to detect if converting dates. Default\n",
    "  None. By default the timestamp precision will be detected, if this is not desired\n",
    "  then pass one of ‘s’, ‘ms’, ‘us’ or ‘ns’ to force timestamp precision to\n",
    "  seconds, milliseconds, microseconds or nanoseconds respectively.  \n",
    "- `lines` : reads file as one json object per line.  \n",
    "- `encoding` : The encoding to use to decode py3 bytes.  \n",
    "- `chunksize` : when used in combination with `lines=True`, return a JsonReader which reads in `chunksize` lines per iteration.  \n",
    "\n",
    "\n",
    "The parser will raise one of `ValueError/TypeError/AssertionError` if the JSON is not parseable.\n",
    "\n",
    "If a non-default `orient` was used when encoding to JSON be sure to pass the same\n",
    "option here so that decoding produces sensible results, see [Orient Options](#orient-options) for an\n",
    "overview."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Data conversion\n",
    "\n",
    "The default of `convert_axes=True`, `dtype=True`, and `convert_dates=True`\n",
    "will try to parse the axes, and all of the data into appropriate types,\n",
    "including dates. If you need to override specific dtypes, pass a dict to\n",
    "`dtype`. `convert_axes` should only be set to `False` if you need to\n",
    "preserve string-like numbers (e.g. ‘1’, ‘2’) in an axes.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">Large integer values may be converted to dates if `convert_dates=True` and the data and / or column labels appear ‘date-like’. The exact threshold depends on the `date_unit` specified. ‘date-like’ means that the column label meets one of the following criteria:\n",
    "\n",
    "> - it ends with `'_at'`  \n",
    "- it ends with `'_time'`  \n",
    "- it begins with `'timestamp'`  \n",
    "- it is `'modified'`  \n",
    "- it is `'date'`  \n",
    "\n",
    "\n",
    "\n",
    "When reading JSON data, automatic coercing into dtypes has some quirks:\n",
    "\n",
    "> - an index can be reconstructed in a different order from serialization, that is, the returned order is not guaranteed to be the same as before serialization  \n",
    "- a column that was `float` data will be converted to `integer` if it can be done safely, e.g. a column of `1.`  \n",
    "- bool columns will be converted to `integer` on reconstruction  \n",
    "\n",
    "\n",
    "\n",
    "Thus there are times where you may want to specify specific dtypes via the `dtype` keyword argument.\n",
    "\n",
    "Reading from a JSON string:\n",
    "\n",
    "Reading from a file:\n",
    "\n",
    "Don’t convert any data (but still convert axes and dates):\n",
    "\n",
    "Specify dtypes for conversion:\n",
    "\n",
    "Preserve string indices:\n",
    "\n",
    "Dates written in nanoseconds need to be read back in nanoseconds:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The Numpy parameter\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">This supports numeric data only. Index and columns labels may be non-numeric, e.g. strings, dates etc.\n",
    "\n",
    "If `numpy=True` is passed to `read_json` an attempt will be made to sniff\n",
    "an appropriate dtype during deserialization and to subsequently decode directly\n",
    "to NumPy arrays, bypassing the need for intermediate Python objects.\n",
    "\n",
    "This can provide speedups if you are deserialising a large amount of numeric\n",
    "data:\n",
    "\n",
    "The speedup is less noticeable for smaller datasets:\n",
    "\n",
    "Direct NumPy decoding makes a number of assumptions and may fail or produce\n",
    "unexpected output if these assumptions are not satisfied:\n",
    "\n",
    "> - data is numeric.  \n",
    "- data is uniform. The dtype is sniffed from the first value decoded.\n",
    "  A `ValueError` may be raised, or incorrect output may be produced\n",
    "  if this condition is not satisfied.  \n",
    "- labels are ordered. Labels are only read from the first container, it is assumed\n",
    "  that each subsequent row / column has been encoded in the same order. This should be satisfied if the\n",
    "  data was encoded using `to_json` but may not be the case if the JSON\n",
    "  is from another source.  \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "<a id='io-json-normalize'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Normalization\n",
    "\n",
    "pandas provides a utility function to take a dict or list of dicts and *normalize* this semi-structured data\n",
    "into a flat table.\n",
    "\n",
    "The max_level parameter provides more control over which level to end normalization.\n",
    "With max_level=1 the following snippet normalizes until 1st nesting level of the provided dict.\n",
    "\n",
    "\n",
    "<a id='io-jsonl'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Line delimited json\n",
    "\n",
    "New in version 0.19.0.\n",
    "\n",
    "pandas is able to read and write line-delimited json files that are common in data processing pipelines\n",
    "using Hadoop or Spark.\n",
    "\n",
    "New in version 0.21.0.\n",
    "\n",
    "For line-delimited json files, pandas can also return an iterator which reads in `chunksize` lines at a time. This can be useful for large files or to read from a stream.\n",
    "\n",
    "\n",
    "<a id='io-table-schema'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Table schema\n",
    "\n",
    "New in version 0.20.0.\n",
    "\n",
    "[Table Schema](https://specs.frictionlessdata.io/json-table-schema/) is a spec for describing tabular datasets as a JSON\n",
    "object. The JSON includes information on the field names, types, and\n",
    "other attributes. You can use the orient `table` to build\n",
    "a JSON string with two fields, `schema` and `data`.\n",
    "\n",
    "The `schema` field contains the `fields` key, which itself contains\n",
    "a list of column name to type pairs, including the `Index` or `MultiIndex`\n",
    "(see below for a list of types).\n",
    "The `schema` field also contains a `primaryKey` field if the (Multi)index\n",
    "is unique.\n",
    "\n",
    "The second field, `data`, contains the serialized data with the `records`\n",
    "orient.\n",
    "The index is included, and any datetimes are ISO 8601 formatted, as required\n",
    "by the Table Schema spec.\n",
    "\n",
    "The full list of types supported are described in the Table Schema\n",
    "spec. This table shows the mapping from pandas types:\n",
    "\n",
    "|Pandas type|Table Schema type|\n",
    "|:-------------:|:---------------:|\n",
    "|int64|integer|\n",
    "|float64|number|\n",
    "|bool|boolean|\n",
    "|datetime64[ns]|datetime|\n",
    "|timedelta64[ns]|duration|\n",
    "|categorical|any|\n",
    "|object|str|\n",
    "A few notes on the generated table schema:\n",
    "\n",
    "- The `schema` object contains a `pandas_version` field. This contains\n",
    "  the version of pandas’ dialect of the schema, and will be incremented\n",
    "  with each revision.  \n",
    "- All dates are converted to UTC when serializing. Even timezone naive values,\n",
    "  which are treated as UTC with an offset of 0.  \n",
    "- datetimes with a timezone (before serializing), include an additional field\n",
    "  `tz` with the time zone name (e.g. `'US/Central'`).  \n",
    "- Periods are converted to timestamps before serialization, and so have the\n",
    "  same behavior of being converted to UTC. In addition, periods will contain\n",
    "  and additional field `freq` with the period’s frequency, e.g. `'A-DEC'`.  \n",
    "- Categoricals use the `any` type and an `enum` constraint listing\n",
    "  the set of possible values. Additionally, an `ordered` field is included:  \n",
    "- A `primaryKey` field, containing an array of labels, is included\n",
    "  *if the index is unique*:  \n",
    "- The `primaryKey` behavior is the same with MultiIndexes, but in this\n",
    "  case the `primaryKey` is an array:  \n",
    "- The default naming roughly follows these rules:  \n",
    "  \n",
    "  - For series, the `object.name` is used. If that’s none, then the\n",
    "    name is `values`  \n",
    "  - For `DataFrames`, the stringified version of the column name is used  \n",
    "  - For `Index` (not `MultiIndex`), `index.name` is used, with a\n",
    "    fallback to `index` if that is None.  \n",
    "  - For `MultiIndex`, `mi.names` is used. If any level has no name,\n",
    "    then `level_<i>` is used.  \n",
    "  \n",
    "\n",
    "\n",
    "New in version 0.23.0.\n",
    "\n",
    "`read_json` also accepts `orient='table'` as an argument. This allows for\n",
    "the preservation of metadata such as dtypes and index names in a\n",
    "round-trippable manner.\n",
    "\n",
    "> \n",
    "Please note that the literal string ‘index’ as the name of an `Index`\n",
    "is not round-trippable, nor are any names beginning with `'level_'` within a\n",
    "`MultiIndex`. These are used by default in `DataFrame.to_json()` to\n",
    "indicate missing values and the subsequent read cannot distinguish the intent."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HTML\n",
    "\n",
    "\n",
    "<a id='io-read-html'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading HTML content\n",
    "\n",
    "We **highly encourage** you to read the [HTML Table Parsing gotchas](#io-html-gotchas)\n",
    "below regarding the issues surrounding the BeautifulSoup4/html5lib/lxml parsers.\n",
    "\n",
    "The top-level `read_html()` function can accept an HTML\n",
    "string/file/URL and will parse HTML tables into list of pandas `DataFrames`.\n",
    "Let’s look at a few examples.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">`read_html` returns a `list` of `DataFrame` objects, even if there is\n",
    "only a single table contained in the HTML content.\n",
    "\n",
    "Read a URL with no options:\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">The data from the above URL changes every Monday so the resulting data above\n",
    "and the data below may be slightly different.\n",
    "\n",
    "Read in the content of the file from the above URL and pass it to `read_html`\n",
    "as a string:\n",
    "\n",
    "You can even pass in an instance of `StringIO` if you so desire:\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">The following examples are not run by the IPython evaluator due to the fact\n",
    "that having so many network-accessing functions slows down the documentation\n",
    "build. If you spot an error or an example that doesn’t run, please do not\n",
    "hesitate to report it over on [pandas GitHub issues page](https://www.github.com/pandas-dev/pandas/issues).\n",
    "\n",
    "Read a URL and match a table that contains specific text:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "match = 'Metcalf Bank'\n",
    "df_list = pd.read_html(url, match=match)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify a header row (by default `<th>` or `<td>` elements located within a\n",
    "`<thead>` are used to form the column index, if multiple rows are contained within\n",
    "`<thead>` then a MultiIndex is created); if specified, the header row is taken\n",
    "from the data minus the parsed header elements (`<th>` elements)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "dfs = pd.read_html(url, header=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify an index column:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "dfs = pd.read_html(url, index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify a number of rows to skip:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "dfs = pd.read_html(url, skiprows=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify a number of rows to skip using a list (`xrange` (Python 2 only) works\n",
    "as well):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "dfs = pd.read_html(url, skiprows=range(2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify an HTML attribute:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "dfs1 = pd.read_html(url, attrs={'id': 'table'})\n",
    "dfs2 = pd.read_html(url, attrs={'class': 'sortable'})\n",
    "print(np.array_equal(dfs1[0], dfs2[0]))  # Should be True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify values that should be converted to NaN:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "dfs = pd.read_html(url, na_values=['No Acquirer'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "New in version 0.19.\n",
    "\n",
    "Specify whether to keep the default set of NaN values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "dfs = pd.read_html(url, keep_default_na=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "New in version 0.19.\n",
    "\n",
    "Specify converters for columns. This is useful for numerical text data that has\n",
    "leading zeros.  By default columns that are numerical are cast to numeric\n",
    "types and the leading zeros are lost. To avoid this, we can convert these\n",
    "columns to strings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "url_mcc = 'https://en.wikipedia.org/wiki/Mobile_country_code'\n",
    "dfs = pd.read_html(url_mcc, match='Telekom Albania', header=0,\n",
    "                   converters={'MNC': str})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "New in version 0.19.\n",
    "\n",
    "Use some combination of the above:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "dfs = pd.read_html(url, match='Metcalf Bank', index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in pandas `to_html` output (with some loss of floating point precision):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "df = pd.DataFrame(np.random.randn(2, 2))\n",
    "s = df.to_html(float_format='{0:.40g}'.format)\n",
    "dfin = pd.read_html(s, index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `lxml` backend will raise an error on a failed parse if that is the only\n",
    "parser you provide. If you only have a single parser you can provide just a\n",
    "string, but it is considered good practice to pass a list with one string if,\n",
    "for example, the function expects a sequence of strings. You may use:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "dfs = pd.read_html(url, 'Metcalf Bank', index_col=0, flavor=['lxml'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or you could pass `flavor='lxml'` without a list:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "dfs = pd.read_html(url, 'Metcalf Bank', index_col=0, flavor='lxml')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "However, if you have bs4 and html5lib installed and pass `None` or `['lxml',\n",
    "'bs4']` then the parse will most likely succeed. Note that *as soon as a parse\n",
    "succeeds, the function will return*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "dfs = pd.read_html(url, 'Metcalf Bank', index_col=0, flavor=['lxml', 'bs4'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<a id='io-html'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing to HTML files\n",
    "\n",
    "`DataFrame` objects have an instance method `to_html` which renders the\n",
    "contents of the `DataFrame` as an HTML table. The function arguments are as\n",
    "in the method `to_string` described above.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">Not all of the possible options for `DataFrame.to_html` are shown here for\n",
    "brevity’s sake. See `to_html()` for the\n",
    "full set of options.\n",
    "\n",
    "HTML:\n",
    "\n",
    "<table border=\"1\" class=\"dataframe\">\n",
    "  <thead>\n",
    "    <tr style=\"text-align: right;\">\n",
    "      <th></th>\n",
    "      <th>0</th>\n",
    "      <th>1</th>\n",
    "    </tr>\n",
    "  </thead>\n",
    "  <tbody>\n",
    "    <tr>\n",
    "      <th>0</th>\n",
    "      <td>-0.184744</td>\n",
    "      <td>0.496971</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "      <th>1</th>\n",
    "      <td>-0.856240</td>\n",
    "      <td>1.857977</td>\n",
    "    </tr>\n",
    "  </tbody>\n",
    "</table>\n",
    "\n",
    "The `columns` argument will limit the columns shown:\n",
    "\n",
    "HTML:\n",
    "\n",
    "<table border=\"1\" class=\"dataframe\">\n",
    "  <thead>\n",
    "    <tr style=\"text-align: right;\">\n",
    "      <th></th>\n",
    "      <th>0</th>\n",
    "    </tr>\n",
    "  </thead>\n",
    "  <tbody>\n",
    "    <tr>\n",
    "      <th>0</th>\n",
    "      <td>-0.184744</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "      <th>1</th>\n",
    "      <td>-0.856240</td>\n",
    "    </tr>\n",
    "  </tbody>\n",
    "</table>\n",
    "\n",
    "`float_format` takes a Python callable to control the precision of floating\n",
    "point values:\n",
    "\n",
    "HTML:\n",
    "\n",
    "<table border=\"1\" class=\"dataframe\">\n",
    "  <thead>\n",
    "    <tr style=\"text-align: right;\">\n",
    "      <th></th>\n",
    "      <th>0</th>\n",
    "      <th>1</th>\n",
    "    </tr>\n",
    "  </thead>\n",
    "  <tbody>\n",
    "    <tr>\n",
    "      <th>0</th>\n",
    "      <td>-0.1847438576</td>\n",
    "      <td>0.4969711327</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "      <th>1</th>\n",
    "      <td>-0.8562396763</td>\n",
    "      <td>1.8579766508</td>\n",
    "    </tr>\n",
    "  </tbody>\n",
    "</table>\n",
    "\n",
    "`bold_rows` will make the row labels bold by default, but you can turn that\n",
    "off:\n",
    "\n",
    "<table border=\"1\" class=\"dataframe\">\n",
    "  <thead>\n",
    "    <tr style=\"text-align: right;\">\n",
    "      <th></th>\n",
    "      <th>0</th>\n",
    "      <th>1</th>\n",
    "    </tr>\n",
    "  </thead>\n",
    "  <tbody>\n",
    "    <tr>\n",
    "      <td>0</td>\n",
    "      <td>-0.184744</td>\n",
    "      <td>0.496971</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "      <td>1</td>\n",
    "      <td>-0.856240</td>\n",
    "      <td>1.857977</td>\n",
    "    </tr>\n",
    "  </tbody>\n",
    "</table>\n",
    "\n",
    "The `classes` argument provides the ability to give the resulting HTML\n",
    "table CSS classes. Note that these classes are *appended* to the existing\n",
    "`'dataframe'` class.\n",
    "\n",
    "The `render_links` argument provides the ability to add hyperlinks to cells\n",
    "that contain URLs.\n",
    "\n",
    "New in version 0.24.\n",
    "\n",
    "HTML:\n",
    "\n",
    "<table border=\"1\" class=\"dataframe\">\n",
    "  <thead>\n",
    "    <tr style=\"text-align: right;\">\n",
    "      <th></th>\n",
    "      <th>name</th>\n",
    "      <th>url</th>\n",
    "    </tr>\n",
    "  </thead>\n",
    "  <tbody>\n",
    "    <tr>\n",
    "      <th>0</th>\n",
    "      <td>Python</td>\n",
    "      <td><a href=\"https://www.python.org/\" target=\"_blank\">https://www.python.org/</a></td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "      <th>1</th>\n",
    "      <td>Pandas</td>\n",
    "      <td><a href=\"http://pandas.pydata.org\" target=\"_blank\">http://pandas.pydata.org</a></td>\n",
    "    </tr>\n",
    "  </tbody>\n",
    "</table>\n",
    "\n",
    "Finally, the `escape` argument allows you to control whether the\n",
    "“<”, “>” and “&” characters escaped in the resulting HTML (by default it is\n",
    "`True`). So to get the HTML without escaped characters pass `escape=False`\n",
    "\n",
    "Escaped:\n",
    "\n",
    "<table border=\"1\" class=\"dataframe\">\n",
    "  <thead>\n",
    "    <tr style=\"text-align: right;\">\n",
    "      <th></th>\n",
    "      <th>a</th>\n",
    "      <th>b</th>\n",
    "    </tr>\n",
    "  </thead>\n",
    "  <tbody>\n",
    "    <tr>\n",
    "      <th>0</th>\n",
    "      <td>&amp;</td>\n",
    "      <td>-0.474063</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "      <th>1</th>\n",
    "      <td>&lt;</td>\n",
    "      <td>-0.230305</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "      <th>2</th>\n",
    "      <td>&gt;</td>\n",
    "      <td>-0.400654</td>\n",
    "    </tr>\n",
    "  </tbody>\n",
    "</table>\n",
    "\n",
    "Not escaped:\n",
    "\n",
    "<table border=\"1\" class=\"dataframe\">\n",
    "  <thead>\n",
    "    <tr style=\"text-align: right;\">\n",
    "      <th></th>\n",
    "      <th>a</th>\n",
    "      <th>b</th>\n",
    "    </tr>\n",
    "  </thead>\n",
    "  <tbody>\n",
    "    <tr>\n",
    "      <th>0</th>\n",
    "      <td>&</td>\n",
    "      <td>-0.474063</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "      <th>1</th>\n",
    "      <td><</td>\n",
    "      <td>-0.230305</td>\n",
    "    </tr>\n",
    "    <tr>\n",
    "      <th>2</th>\n",
    "      <td>></td>\n",
    "      <td>-0.400654</td>\n",
    "    </tr>\n",
    "  </tbody>\n",
    "</table>\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">Some browsers may not show a difference in the rendering of the previous two\n",
    "HTML tables.\n",
    "\n",
    "\n",
    "<a id='io-html-gotchas'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### HTML Table Parsing Gotchas\n",
    "\n",
    "There are some versioning issues surrounding the libraries that are used to\n",
    "parse HTML tables in the top-level pandas io function `read_html`.\n",
    "\n",
    "**Issues with** [**lxml**](https://lxml.de)\n",
    "\n",
    "- Benefits  \n",
    "  \n",
    "  - [**lxml**](https://lxml.de) is very fast.  \n",
    "  - [**lxml**](https://lxml.de) requires Cython to install correctly.  \n",
    "  \n",
    "- Drawbacks  \n",
    "  \n",
    "  - [**lxml**](https://lxml.de) does *not* make any guarantees about the results of its parse\n",
    "    *unless* it is given [**strictly valid markup**](https://validator.w3.org/docs/help.html#validation_basics).  \n",
    "  - In light of the above, we have chosen to allow you, the user, to use the\n",
    "    [**lxml**](https://lxml.de) backend, but **this backend will use** [**html5lib**](https://github.com/html5lib/html5lib-python) if [**lxml**](https://lxml.de)\n",
    "    fails to parse  \n",
    "  - It is therefore *highly recommended* that you install both\n",
    "    [**BeautifulSoup4**](https://www.crummy.com/software/BeautifulSoup) and [**html5lib**](https://github.com/html5lib/html5lib-python), so that you will still get a valid\n",
    "    result (provided everything else is valid) even if [**lxml**](https://lxml.de) fails.  \n",
    "  \n",
    "\n",
    "\n",
    "**Issues with** [**BeautifulSoup4**](https://www.crummy.com/software/BeautifulSoup) **using** [**lxml**](https://lxml.de) **as a backend**\n",
    "\n",
    "- The above issues hold here as well since [**BeautifulSoup4**](https://www.crummy.com/software/BeautifulSoup) is essentially\n",
    "  just a wrapper around a parser backend.  \n",
    "\n",
    "\n",
    "**Issues with** [**BeautifulSoup4**](https://www.crummy.com/software/BeautifulSoup) **using** [**html5lib**](https://github.com/html5lib/html5lib-python) **as a backend**\n",
    "\n",
    "- Benefits  \n",
    "  \n",
    "  - [**html5lib**](https://github.com/html5lib/html5lib-python) is far more lenient than [**lxml**](https://lxml.de) and consequently deals\n",
    "    with *real-life markup* in a much saner way rather than just, e.g.,\n",
    "    dropping an element without notifying you.  \n",
    "  - [**html5lib**](https://github.com/html5lib/html5lib-python) *generates valid HTML5 markup from invalid markup\n",
    "    automatically*. This is extremely important for parsing HTML tables,\n",
    "    since it guarantees a valid document. However, that does NOT mean that\n",
    "    it is “correct”, since the process of fixing markup does not have a\n",
    "    single definition.  \n",
    "  - [**html5lib**](https://github.com/html5lib/html5lib-python) is pure Python and requires no additional build steps beyond\n",
    "    its own installation.  \n",
    "  \n",
    "- Drawbacks  \n",
    "  \n",
    "  - The biggest drawback to using [**html5lib**](https://github.com/html5lib/html5lib-python) is that it is slow as\n",
    "    molasses.  However consider the fact that many tables on the web are not\n",
    "    big enough for the parsing algorithm runtime to matter. It is more\n",
    "    likely that the bottleneck will be in the process of reading the raw\n",
    "    text from the URL over the web, i.e., IO (input-output). For very large\n",
    "    tables, this might not be true.  \n",
    "  \n",
    "\n",
    "\n",
    "\n",
    "<a id='io-excel'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Excel files\n",
    "\n",
    "The `read_excel()` method can read Excel 2003 (`.xls`)\n",
    "files using the `xlrd` Python module.  Excel 2007+ (`.xlsx`) files\n",
    "can be read using either `xlrd` or `openpyxl`.\n",
    "The `to_excel()` instance method is used for\n",
    "saving a `DataFrame` to Excel.  Generally the semantics are\n",
    "similar to working with [csv](#io-read-csv-table) data.\n",
    "See the [cookbook](cookbook.ipynb#cookbook-excel) for some advanced strategies.\n",
    "\n",
    "\n",
    "<a id='io-excel-reader'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading Excel files\n",
    "\n",
    "In the most basic use-case, `read_excel` takes a path to an Excel\n",
    "file, and the `sheet_name` indicating which sheet to parse."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "# Returns a DataFrame\n",
    "pd.read_excel('path_to_file.xls', sheet_name='Sheet1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<a id='io-excel-excelfile-class'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### `ExcelFile` class\n",
    "\n",
    "To facilitate working with multiple sheets from the same file, the `ExcelFile`\n",
    "class can be used to wrap the file and can be passed into `read_excel`\n",
    "There will be a performance benefit for reading multiple sheets as the file is\n",
    "read into memory only once."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "xlsx = pd.ExcelFile('path_to_file.xls')\n",
    "df = pd.read_excel(xlsx, 'Sheet1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `ExcelFile` class can also be used as a context manager."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "with pd.ExcelFile('path_to_file.xls') as xls:\n",
    "    df1 = pd.read_excel(xls, 'Sheet1')\n",
    "    df2 = pd.read_excel(xls, 'Sheet2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `sheet_names` property will generate\n",
    "a list of the sheet names in the file.\n",
    "\n",
    "The primary use-case for an `ExcelFile` is parsing multiple sheets with\n",
    "different parameters:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "data = {}\n",
    "# For when Sheet1's format differs from Sheet2\n",
    "with pd.ExcelFile('path_to_file.xls') as xls:\n",
    "    data['Sheet1'] = pd.read_excel(xls, 'Sheet1', index_col=None,\n",
    "                                   na_values=['NA'])\n",
    "    data['Sheet2'] = pd.read_excel(xls, 'Sheet2', index_col=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that if the same parsing parameters are used for all sheets, a list\n",
    "of sheet names can simply be passed to `read_excel` with no loss in performance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "# using the ExcelFile class\n",
    "data = {}\n",
    "with pd.ExcelFile('path_to_file.xls') as xls:\n",
    "    data['Sheet1'] = pd.read_excel(xls, 'Sheet1', index_col=None,\n",
    "                                   na_values=['NA'])\n",
    "    data['Sheet2'] = pd.read_excel(xls, 'Sheet2', index_col=None,\n",
    "                                   na_values=['NA'])\n",
    "\n",
    "# equivalent using the read_excel function\n",
    "data = pd.read_excel('path_to_file.xls', ['Sheet1', 'Sheet2'],\n",
    "                     index_col=None, na_values=['NA'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`ExcelFile` can also be called with a `xlrd.book.Book` object\n",
    "as a parameter. This allows the user to control how the excel file is read.\n",
    "For example, sheets can be loaded on demand by calling `xlrd.open_workbook()`\n",
    "with `on_demand=True`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "import xlrd\n",
    "xlrd_book = xlrd.open_workbook('path_to_file.xls', on_demand=True)\n",
    "with pd.ExcelFile(xlrd_book) as xls:\n",
    "    df1 = pd.read_excel(xls, 'Sheet1')\n",
    "    df2 = pd.read_excel(xls, 'Sheet2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<a id='io-excel-specifying-sheets'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Specifying sheets\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">The second argument is `sheet_name`, not to be confused with `ExcelFile.sheet_names`.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">An ExcelFile’s attribute `sheet_names` provides access to a list of sheets.\n",
    "\n",
    "- The arguments `sheet_name` allows specifying the sheet or sheets to read.  \n",
    "- The default value for `sheet_name` is 0, indicating to read the first sheet  \n",
    "- Pass a string to refer to the name of a particular sheet in the workbook.  \n",
    "- Pass an integer to refer to the index of a sheet. Indices follow Python\n",
    "  convention, beginning at 0.  \n",
    "- Pass a list of either strings or integers, to return a dictionary of specified sheets.  \n",
    "- Pass a `None` to return a dictionary of all available sheets.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "# Returns a DataFrame\n",
    "pd.read_excel('path_to_file.xls', 'Sheet1', index_col=None, na_values=['NA'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the sheet index:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "# Returns a DataFrame\n",
    "pd.read_excel('path_to_file.xls', 0, index_col=None, na_values=['NA'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using all default values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "# Returns a DataFrame\n",
    "pd.read_excel('path_to_file.xls')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using None to get all sheets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "# Returns a dictionary of DataFrames\n",
    "pd.read_excel('path_to_file.xls', sheet_name=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using a list to get multiple sheets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "# Returns the 1st and 4th sheet, as a dictionary of DataFrames.\n",
    "pd.read_excel('path_to_file.xls', sheet_name=['Sheet1', 3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`read_excel` can read more than one sheet, by setting `sheet_name` to either\n",
    "a list of sheet names, a list of sheet positions, or `None` to read all sheets.\n",
    "Sheets can be specified by sheet index or sheet name, using an integer or string,\n",
    "respectively.\n",
    "\n",
    "\n",
    "<a id='io-excel-reading-multiindex'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reading a `MultiIndex`\n",
    "\n",
    "`read_excel` can read a `MultiIndex` index, by passing a list of columns to `index_col`\n",
    "and a `MultiIndex` column by passing a list of rows to `header`.  If either the `index`\n",
    "or `columns` have serialized level names those will be read in as well by specifying\n",
    "the rows/columns that make up the levels.\n",
    "\n",
    "For example, to read in a `MultiIndex` index without names:\n",
    "\n",
    "If the index has level names, they will parsed as well, using the same\n",
    "parameters.\n",
    "\n",
    "If the source file has both `MultiIndex` index and columns, lists specifying each\n",
    "should be passed to `index_col` and `header`:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Parsing specific columns\n",
    "\n",
    "It is often the case that users will insert columns to do temporary computations\n",
    "in Excel and you may not want to read in those columns. `read_excel` takes\n",
    "a `usecols` keyword to allow you to specify a subset of columns to parse.\n",
    "\n",
    "Deprecated since version 0.24.0.\n",
    "\n",
    "Passing in an integer for `usecols` has been deprecated. Please pass in a list\n",
    "of ints from 0 to `usecols` inclusive instead.\n",
    "\n",
    "If `usecols` is an integer, then it is assumed to indicate the last column\n",
    "to be parsed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "pd.read_excel('path_to_file.xls', 'Sheet1', usecols=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can also specify a comma-delimited set of Excel columns and ranges as a string:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "pd.read_excel('path_to_file.xls', 'Sheet1', usecols='A,C:E')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If `usecols` is a list of integers, then it is assumed to be the file column\n",
    "indices to be parsed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "pd.read_excel('path_to_file.xls', 'Sheet1', usecols=[0, 2, 3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Element order is ignored, so `usecols=[0, 1]` is the same as `[1, 0]`.\n",
    "\n",
    "New in version 0.24.\n",
    "\n",
    "If `usecols` is a list of strings, it is assumed that each string corresponds\n",
    "to a column name provided either by the user in `names` or inferred from the\n",
    "document header row(s). Those strings define which columns will be parsed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "pd.read_excel('path_to_file.xls', 'Sheet1', usecols=['foo', 'bar'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Element order is ignored, so `usecols=['baz', 'joe']` is the same as `['joe', 'baz']`.\n",
    "\n",
    "New in version 0.24.\n",
    "\n",
    "If `usecols` is callable, the callable function will be evaluated against\n",
    "the column names, returning names where the callable function evaluates to `True`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "pd.read_excel('path_to_file.xls', 'Sheet1', usecols=lambda x: x.isalpha())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Parsing dates\n",
    "\n",
    "Datetime-like values are normally automatically converted to the appropriate\n",
    "dtype when reading the excel file. But if you have a column of strings that\n",
    "*look* like dates (but are not actually formatted as dates in excel), you can\n",
    "use the `parse_dates` keyword to parse those strings to datetimes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "pd.read_excel('path_to_file.xls', 'Sheet1', parse_dates=['date_strings'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cell converters\n",
    "\n",
    "It is possible to transform the contents of Excel cells via the `converters`\n",
    "option. For instance, to convert a column to boolean:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "pd.read_excel('path_to_file.xls', 'Sheet1', converters={'MyBools': bool})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This options handles missing values and treats exceptions in the converters\n",
    "as missing data. Transformations are applied cell by cell rather than to the\n",
    "column as a whole, so the array dtype is not guaranteed. For instance, a\n",
    "column of integers with missing values cannot be transformed to an array\n",
    "with integer dtype, because NaN is strictly a float. You can manually mask\n",
    "missing data to recover integer dtype:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "def cfun(x):\n",
    "    return int(x) if x else -1\n",
    "\n",
    "\n",
    "pd.read_excel('path_to_file.xls', 'Sheet1', converters={'MyInts': cfun})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Dtype specifications\n",
    "\n",
    "New in version 0.20.\n",
    "\n",
    "As an alternative to converters, the type for an entire column can\n",
    "be specified using the dtype keyword, which takes a dictionary\n",
    "mapping column names to types.  To interpret data with\n",
    "no type inference, use the type `str` or `object`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "pd.read_excel('path_to_file.xls', dtype={'MyInts': 'int64', 'MyText': str})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<a id='io-excel-writer'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing Excel files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Writing Excel files to disk\n",
    "\n",
    "To write a `DataFrame` object to a sheet of an Excel file, you can use the\n",
    "`to_excel` instance method.  The arguments are largely the same as `to_csv`\n",
    "described above, the first argument being the name of the excel file, and the\n",
    "optional second argument the name of the sheet to which the `DataFrame` should be\n",
    "written. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "df.to_excel('path_to_file.xlsx', sheet_name='Sheet1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Files with a `.xls` extension will be written using `xlwt` and those with a\n",
    "`.xlsx` extension will be written using `xlsxwriter` (if available) or\n",
    "`openpyxl`.\n",
    "\n",
    "The `DataFrame` will be written in a way that tries to mimic the REPL output.\n",
    "The `index_label` will be placed in the second\n",
    "row instead of the first. You can place it in the first row by setting the\n",
    "`merge_cells` option in `to_excel()` to `False`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "df.to_excel('path_to_file.xlsx', index_label='label', merge_cells=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to write separate `DataFrames` to separate sheets in a single Excel file,\n",
    "one can pass an `ExcelWriter`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "with pd.ExcelWriter('path_to_file.xlsx') as writer:\n",
    "    df1.to_excel(writer, sheet_name='Sheet1')\n",
    "    df2.to_excel(writer, sheet_name='Sheet2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ">**Note**\n",
    ">\n",
    ">Wringing a little more performance out of `read_excel`\n",
    "Internally, Excel stores all numeric data as floats. Because this can\n",
    "produce unexpected behavior when reading in data, pandas defaults to trying\n",
    "to convert integers to floats if it doesn’t lose information (`1.0 -->\n",
    "1`).  You can pass `convert_float=False` to disable this behavior, which\n",
    "may give a slight performance improvement.\n",
    "\n",
    "\n",
    "<a id='io-excel-writing-buffer'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Writing Excel files to memory\n",
    "\n",
    "Pandas supports writing Excel files to buffer-like objects such as `StringIO` or\n",
    "`BytesIO` using `ExcelWriter`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "# Safe import for either Python 2.x or 3.x\n",
    "try:\n",
    "    from io import BytesIO\n",
    "except ImportError:\n",
    "    from cStringIO import StringIO as BytesIO\n",
    "\n",
    "bio = BytesIO()\n",
    "\n",
    "# By setting the 'engine' in the ExcelWriter constructor.\n",
    "writer = pd.ExcelWriter(bio, engine='xlsxwriter')\n",
    "df.to_excel(writer, sheet_name='Sheet1')\n",
    "\n",
    "# Save the workbook\n",
    "writer.save()\n",
    "\n",
    "# Seek to the beginning and read to copy the workbook to a variable in memory\n",
    "bio.seek(0)\n",
    "workbook = bio.read()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ">**Note**\n",
    ">\n",
    ">`engine` is optional but recommended.  Setting the engine determines\n",
    "the version of workbook produced. Setting `engine='xlrd'` will produce an\n",
    "Excel 2003-format workbook (xls).  Using either `'openpyxl'` or\n",
    "`'xlsxwriter'` will produce an Excel 2007-format workbook (xlsx). If\n",
    "omitted, an Excel 2007-formatted workbook is produced.\n",
    "\n",
    "\n",
    "<a id='io-excel-writers'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Excel writer engines\n",
    "\n",
    "Pandas chooses an Excel writer via two methods:\n",
    "\n",
    "1. the `engine` keyword argument  \n",
    "1. the filename extension (via the default specified in config options)  \n",
    "\n",
    "\n",
    "By default, pandas uses the [XlsxWriter](https://xlsxwriter.readthedocs.io)  for `.xlsx`, [openpyxl](https://openpyxl.readthedocs.io/)\n",
    "for `.xlsm`, and [xlwt](http://www.python-excel.org) for `.xls` files. If you have multiple\n",
    "engines installed, you can set the default engine through [setting the\n",
    "config options](options.ipynb#options) `io.excel.xlsx.writer` and\n",
    "`io.excel.xls.writer`. pandas will fall back on [openpyxl](https://openpyxl.readthedocs.io/) for `.xlsx`\n",
    "files if [Xlsxwriter](https://xlsxwriter.readthedocs.io) is not available.\n",
    "\n",
    "To specify which writer you want to use, you can pass an engine keyword\n",
    "argument to `to_excel` and to `ExcelWriter`. The built-in engines are:\n",
    "\n",
    "- `openpyxl`: version 2.4 or higher is required  \n",
    "- `xlsxwriter`  \n",
    "- `xlwt`  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "# By setting the 'engine' in the DataFrame 'to_excel()' methods.\n",
    "df.to_excel('path_to_file.xlsx', sheet_name='Sheet1', engine='xlsxwriter')\n",
    "\n",
    "# By setting the 'engine' in the ExcelWriter constructor.\n",
    "writer = pd.ExcelWriter('path_to_file.xlsx', engine='xlsxwriter')\n",
    "\n",
    "# Or via pandas configuration.\n",
    "from pandas import options                                     # noqa: E402\n",
    "options.io.excel.xlsx.writer = 'xlsxwriter'\n",
    "\n",
    "df.to_excel('path_to_file.xlsx', sheet_name='Sheet1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<a id='io-excel-style'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Style and formatting\n",
    "\n",
    "The look and feel of Excel worksheets created from pandas can be modified using the following parameters on the `DataFrame`’s `to_excel` method.\n",
    "\n",
    "- `float_format` : Format string for floating point numbers (default `None`).  \n",
    "- `freeze_panes` : A tuple of two integers representing the bottommost row and rightmost column to freeze. Each of these parameters is one-based, so (1, 1) will freeze the first row and first column (default `None`).  \n",
    "\n",
    "\n",
    "Using the [Xlsxwriter](https://xlsxwriter.readthedocs.io) engine provides many options for controlling the\n",
    "format of an Excel worksheet created with the `to_excel` method.  Excellent examples can be found in the\n",
    "[Xlsxwriter](https://xlsxwriter.readthedocs.io) documentation here: [https://xlsxwriter.readthedocs.io/working_with_pandas.html](https://xlsxwriter.readthedocs.io/working_with_pandas.html)\n",
    "\n",
    "\n",
    "<a id='io-ods'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## OpenDocument Spreadsheets\n",
    "\n",
    "New in version 0.25.\n",
    "\n",
    "The `read_excel()` method can also read OpenDocument spreadsheets\n",
    "using the `odfpy` module. The semantics and features for reading\n",
    "OpenDocument spreadsheets match what can be done for [Excel files](#excel-files) using\n",
    "`engine='odf'`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "# Returns a DataFrame\n",
    "pd.read_excel('path_to_file.ods', engine='odf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ">**Note**\n",
    ">\n",
    ">Currently pandas only supports *reading* OpenDocument spreadsheets. Writing\n",
    "is not implemented.\n",
    "\n",
    "\n",
    "<a id='io-clipboard'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clipboard\n",
    "\n",
    "A handy way to grab data is to use the `read_clipboard()` method,\n",
    "which takes the contents of the clipboard buffer and passes them to the\n",
    "`read_csv` method. For instance, you can copy the following text to the\n",
    "clipboard (CTRL-C on many operating systems):"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hide-output": false
   },
   "source": [
    "```console\n",
    "  A B C\n",
    "x 1 4 p\n",
    "y 2 5 q\n",
    "z 3 6 r\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And then import the data directly to a `DataFrame` by calling:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    ">>> clipdf = pd.read_clipboard()\n",
    ">>> clipdf\n",
    "  A B C\n",
    "x 1 4 p\n",
    "y 2 5 q\n",
    "z 3 6 r"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `to_clipboard` method can be used to write the contents of a `DataFrame` to\n",
    "the clipboard. Following which you can paste the clipboard contents into other\n",
    "applications (CTRL-V on many operating systems). Here we illustrate writing a\n",
    "`DataFrame` into clipboard and reading it back."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    ">>> df = pd.DataFrame({'A': [1, 2, 3],\n",
    "...                    'B': [4, 5, 6],\n",
    "...                    'C': ['p', 'q', 'r']},\n",
    "...                   index=['x', 'y', 'z'])\n",
    ">>> df\n",
    "  A B C\n",
    "x 1 4 p\n",
    "y 2 5 q\n",
    "z 3 6 r\n",
    ">>> df.to_clipboard()\n",
    ">>> pd.read_clipboard()\n",
    "  A B C\n",
    "x 1 4 p\n",
    "y 2 5 q\n",
    "z 3 6 r"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that we got the same content back, which we had earlier written to the clipboard.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">You may need to install xclip or xsel (with PyQt5, PyQt4 or qtpy) on Linux to use these methods.\n",
    "\n",
    "\n",
    "<a id='io-pickle'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pickling\n",
    "\n",
    "All pandas objects are equipped with `to_pickle` methods which use Python’s\n",
    "`cPickle` module to save data structures to disk using the pickle format.\n",
    "\n",
    "The `read_pickle` function in the `pandas` namespace can be used to load\n",
    "any pickled pandas object (or any other pickled object) from file:\n",
    "\n",
    "Loading pickled data received from untrusted sources can be unsafe.\n",
    "\n",
    "See: [https://docs.python.org/3/library/pickle.html](https://docs.python.org/3/library/pickle.html)\n",
    "\n",
    "`read_pickle()` is only guaranteed backwards compatible back to pandas version 0.20.3\n",
    "\n",
    "\n",
    "<a id='io-pickle-compression'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compressed pickle files\n",
    "\n",
    "New in version 0.20.0.\n",
    "\n",
    "`read_pickle()`, `DataFrame.to_pickle()` and `Series.to_pickle()` can read\n",
    "and write compressed pickle files. The compression types of `gzip`, `bz2`, `xz` are supported for reading and writing.\n",
    "The `zip` file format only supports reading and must contain only one data file\n",
    "to be read.\n",
    "\n",
    "The compression type can be an explicit parameter or be inferred from the file extension.\n",
    "If ‘infer’, then use `gzip`, `bz2`, `zip`, or `xz` if filename ends in `'.gz'`, `'.bz2'`, `'.zip'`, or\n",
    "`'.xz'`, respectively.\n",
    "\n",
    "Using an explicit compression type:\n",
    "\n",
    "Inferring compression type from the extension:\n",
    "\n",
    "The default is to ‘infer’:\n",
    "\n",
    "\n",
    "<a id='io-msgpack'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## msgpack\n",
    "\n",
    "pandas supports the `msgpack` format for\n",
    "object serialization. This is a lightweight portable binary format, similar\n",
    "to binary JSON, that is highly space efficient, and provides good performance\n",
    "both on the writing (serialization), and reading (deserialization).\n",
    "\n",
    "The msgpack format is deprecated as of 0.25 and will be removed in a future version.\n",
    "It is recommended to use pyarrow for on-the-wire transmission of pandas objects.\n",
    "\n",
    "`read_msgpack()` is only guaranteed backwards compatible back to pandas version 0.20.3\n",
    "\n",
    "You can pass a list of objects and you will receive them back on deserialization.\n",
    "\n",
    "You can pass `iterator=True` to iterate over the unpacked results:\n",
    "\n",
    "You can pass `append=True` to the writer to append to an existing pack:\n",
    "\n",
    "Unlike other io methods, `to_msgpack` is available on both a per-object basis,\n",
    "`df.to_msgpack()` and using the top-level `pd.to_msgpack(...)` where you\n",
    "can pack arbitrary collections of Python lists, dicts, scalars, while intermixing\n",
    "pandas objects."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read/write API\n",
    "\n",
    "Msgpacks can also be read from and written to strings.\n",
    "\n",
    "Furthermore you can concatenate the strings to produce a list of the original objects.\n",
    "\n",
    "\n",
    "<a id='io-hdf5'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HDF5 (PyTables)\n",
    "\n",
    "`HDFStore` is a dict-like object which reads and writes pandas using\n",
    "the high performance HDF5 format using the excellent [PyTables](https://www.pytables.org/) library. See the [cookbook](cookbook.ipynb#cookbook-hdf)\n",
    "for some advanced strategies\n",
    "\n",
    "pandas requires `PyTables` >= 3.0.0.\n",
    "There is a indexing bug in `PyTables` < 3.2 which may appear when querying stores using an index.\n",
    "If you see a subset of results being returned, upgrade to `PyTables` >= 3.2.\n",
    "Stores created previously will need to be rewritten using the updated version.\n",
    "\n",
    "Objects can be written to the file just like adding key-value pairs to a\n",
    "dict:\n",
    "\n",
    "In a current or later Python session, you can retrieve stored objects:\n",
    "\n",
    "Deletion of the object specified by the key:\n",
    "\n",
    "Closing a Store and using a context manager:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read/write API\n",
    "\n",
    "`HDFStore` supports an top-level API using  `read_hdf` for reading and `to_hdf` for writing,\n",
    "similar to how `read_csv` and `to_csv` work.\n",
    "\n",
    "HDFStore will by default not drop rows that are all missing. This behavior can be changed by setting `dropna=True`.\n",
    "\n",
    "\n",
    "<a id='io-hdf5-fixed'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fixed format\n",
    "\n",
    "The examples above show storing using `put`, which write the HDF5 to `PyTables` in a fixed array format, called\n",
    "the `fixed` format. These types of stores are **not** appendable once written (though you can simply\n",
    "remove them and rewrite). Nor are they **queryable**; they must be\n",
    "retrieved in their entirety. They also do not support dataframes with non-unique column names.\n",
    "The `fixed` format stores offer very fast writing and slightly faster reading than `table` stores.\n",
    "This format is specified by default when using `put` or `to_hdf` or by `format='fixed'` or `format='f'`.\n",
    "\n",
    "A `fixed` format will raise a `TypeError` if you try to retrieve using a `where`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    ">>> pd.DataFrame(np.random.randn(10, 2)).to_hdf('test_fixed.h5', 'df')\n",
    ">>> pd.read_hdf('test_fixed.h5', 'df', where='index>5')\n",
    "TypeError: cannot pass a where specification when reading a fixed format.\n",
    "           this store must be selected in its entirety"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<a id='io-hdf5-table'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Table format\n",
    "\n",
    "`HDFStore` supports another `PyTables` format on disk, the `table`\n",
    "format. Conceptually a `table` is shaped very much like a DataFrame,\n",
    "with rows and columns. A `table` may be appended to in the same or\n",
    "other sessions.  In addition, delete and query type operations are\n",
    "supported. This format is specified by `format='table'` or `format='t'`\n",
    "to `append` or `put` or `to_hdf`.\n",
    "\n",
    "This format can be set as an option as well `pd.set_option('io.hdf.default_format','table')` to\n",
    "enable `put/append/to_hdf` to by default store in the `table` format.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">You can also create a `table` by passing `format='table'` or `format='t'` to a `put` operation.\n",
    "\n",
    "\n",
    "<a id='io-hdf5-keys'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Hierarchical keys\n",
    "\n",
    "Keys to a store can be specified as a string. These can be in a\n",
    "hierarchical path-name like format (e.g. `foo/bar/bah`), which will\n",
    "generate a hierarchy of sub-stores (or `Groups` in PyTables\n",
    "parlance). Keys can be specified with out the leading ‘/’ and are **always**\n",
    "absolute (e.g. ‘foo’ refers to ‘/foo’). Removal operations can remove\n",
    "everything in the sub-store and **below**, so be *careful*.\n",
    "\n",
    "You can walk through the group hierarchy using the `walk` method which\n",
    "will yield a tuple for each group key along with the relative keys of its contents.\n",
    "\n",
    "New in version 0.24.0.\n",
    "\n",
    "Hierarchical keys cannot be retrieved as dotted (attribute) access as described above for items stored under the root node."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hide-output": false
   },
   "source": [
    "```ipython\n",
    "In [8]: store.foo.bar.bah\n",
    "AttributeError: 'HDFStore' object has no attribute 'foo'\n",
    "\n",
    "# you can directly access the actual PyTables node but using the root node\n",
    "In [9]: store.root.foo.bar.bah\n",
    "Out[9]:\n",
    "/foo/bar/bah (Group) ''\n",
    "  children := ['block0_items' (Array), 'block0_values' (Array), 'axis0' (Array), 'axis1' (Array)]\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Instead, use explicit string based keys:\n",
    "\n",
    "\n",
    "<a id='io-hdf5-types'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Storing types"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Storing mixed types in a table\n",
    "\n",
    "Storing mixed-dtype data is supported. Strings are stored as a\n",
    "fixed-width using the maximum size of the appended column. Subsequent attempts\n",
    "at appending longer strings will raise a `ValueError`.\n",
    "\n",
    "Passing `min_itemsize={`values`: size}` as a parameter to append\n",
    "will set a larger minimum for the string columns. Storing `floats,\n",
    "strings, ints, bools, datetime64` are currently supported. For string\n",
    "columns, passing `nan_rep = 'nan'` to append will change the default\n",
    "nan representation on disk (which converts to/from np.nan), this\n",
    "defaults to nan."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Storing MultiIndex DataFrames\n",
    "\n",
    "Storing MultiIndex `DataFrames` as tables is very similar to\n",
    "storing/selecting from homogeneous index `DataFrames`.\n",
    "\n",
    "\n",
    "<a id='io-hdf5-query'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Querying"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Querying a table\n",
    "\n",
    "`select` and `delete` operations have an optional criterion that can\n",
    "be specified to select/delete only a subset of the data. This allows one\n",
    "to have a very large on-disk table and retrieve only a portion of the\n",
    "data.\n",
    "\n",
    "A query is specified using the `Term` class under the hood, as a boolean expression.\n",
    "\n",
    "- `index` and `columns` are supported indexers of a `DataFrames`.  \n",
    "- if `data_columns` are specified, these can be used as additional indexers.  \n",
    "\n",
    "\n",
    "Valid comparison operators are:\n",
    "\n",
    "`=, ==, !=, >, >=, <, <=`\n",
    "\n",
    "Valid boolean expressions are combined with:\n",
    "\n",
    "- `|` : or  \n",
    "- `&` : and  \n",
    "- `(` and `)` : for grouping  \n",
    "\n",
    "\n",
    "These rules are similar to how boolean expressions are used in pandas for indexing.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">- `=` will be automatically expanded to the comparison operator `==`  \n",
    "- `~` is the not operator, but can only be used in very limited\n",
    "  circumstances  \n",
    "- If a list/tuple of expressions is passed they will be combined via `&`  \n",
    "\n",
    "\n",
    "The following are valid expressions:\n",
    "\n",
    "- `'index >= date'`  \n",
    "- `\"columns = ['A', 'D']\"`  \n",
    "- `\"columns in ['A', 'D']\"`  \n",
    "- `'columns = A'`  \n",
    "- `'columns == A'`  \n",
    "- `\"~(columns = ['A', 'B'])\"`  \n",
    "- `'index > df.index[3] & string = \"bar\"'`  \n",
    "- `'(index > df.index[3] & index <= df.index[6]) | string = \"bar\"'`  \n",
    "- `\"ts >= Timestamp('2012-02-01')\"`  \n",
    "- `\"major_axis>=20130101\"`  \n",
    "\n",
    "\n",
    "The `indexers` are on the left-hand side of the sub-expression:\n",
    "\n",
    "`columns`, `major_axis`, `ts`\n",
    "\n",
    "The right-hand side of the sub-expression (after a comparison operator) can be:\n",
    "\n",
    "- functions that will be evaluated, e.g. `Timestamp('2012-02-01')`  \n",
    "- strings, e.g. `\"bar\"`  \n",
    "- date-like, e.g. `20130101`, or `\"20130101\"`  \n",
    "- lists, e.g. `\"['A', 'B']\"`  \n",
    "- variables that are defined in the local names space, e.g. `date`  \n",
    "\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">Passing a string to a query by interpolating it into the query\n",
    "expression is not recommended. Simply assign the string of interest to a\n",
    "variable and use that variable in an expression. For example, do this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "string = \"HolyMoly'\"\n",
    "store.select('df', 'index == string')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "instead of this"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hide-output": false
   },
   "source": [
    "```ipython\n",
    "string = \"HolyMoly'\"\n",
    "store.select('df', 'index == %s' % string)\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The latter will **not** work and will raise a `SyntaxError`.Note that\n",
    "there’s a single quote followed by a double quote in the `string`\n",
    "variable.\n",
    "\n",
    "If you *must* interpolate, use the `'%r'` format specifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "store.select('df', 'index == %r' % string)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "which will quote `string`.\n",
    "\n",
    "Here are some examples:\n",
    "\n",
    "Use boolean expressions, with in-line function evaluation.\n",
    "\n",
    "Use and inline column reference\n",
    "\n",
    "The `columns` keyword can be supplied to select a list of columns to be\n",
    "returned, this is equivalent to passing a\n",
    "`'columns=list_of_columns_to_filter'`:\n",
    "\n",
    "`start` and `stop` parameters can be specified to limit the total search\n",
    "space. These are in terms of the total number of rows in a table.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">`select` will raise a `ValueError` if the query expression has an unknown\n",
    "variable reference. Usually this means that you are trying to select on a column\n",
    "that is **not** a data_column.\n",
    "\n",
    "`select` will raise a `SyntaxError` if the query expression is not valid.\n",
    "\n",
    "\n",
    "<a id='io-hdf5-timedelta'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Using timedelta64[ns]\n",
    "\n",
    "You can store and query using the `timedelta64[ns]` type. Terms can be\n",
    "specified in the format: `<float>(<unit>)`, where float may be signed (and fractional), and unit can be\n",
    "`D,s,ms,us,ns` for the timedelta. Here’s an example:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Indexing\n",
    "\n",
    "You can create/modify an index for a table with `create_table_index`\n",
    "after data is already in the table (after and `append/put`\n",
    "operation). Creating a table index is **highly** encouraged. This will\n",
    "speed your queries a great deal when you use a `select` with the\n",
    "indexed dimension as the `where`.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">Indexes are automagically created on the indexables\n",
    "and any data columns you specify. This behavior can be turned off by passing\n",
    "`index=False` to `append`.\n",
    "\n",
    "Oftentimes when appending large amounts of data to a store, it is useful to turn off index creation for each append, then recreate at the end.\n",
    "\n",
    "Then create the index when finished appending.\n",
    "\n",
    "See [here](https://stackoverflow.com/questions/17893370/ptrepack-sortby-needs-full-index) for how to create a completely-sorted-index (CSI) on an existing store.\n",
    "\n",
    "\n",
    "<a id='io-hdf5-query-data-columns'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Query via data columns\n",
    "\n",
    "You can designate (and index) certain columns that you want to be able\n",
    "to perform queries (other than the indexable columns, which you can\n",
    "always query). For instance say you want to perform this common\n",
    "operation, on-disk, and return just the frame that matches this\n",
    "query. You can specify `data_columns = True` to force all columns to\n",
    "be `data_columns`.\n",
    "\n",
    "There is some performance degradation by making lots of columns into\n",
    "data columns, so it is up to the user to designate these. In addition,\n",
    "you cannot change data columns (nor indexables) after the first\n",
    "append/put operation (Of course you can simply read in the data and\n",
    "create a new table!)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Iterator\n",
    "\n",
    "You can pass `iterator=True` or `chunksize=number_in_a_chunk`\n",
    "to `select` and `select_as_multiple` to return an iterator on the results.\n",
    "The default is 50,000 rows returned in a chunk.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">You can also use the iterator with `read_hdf` which will open, then\n",
    "automatically close the store when finished iterating."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "for df in pd.read_hdf('store.h5', 'df', chunksize=3):\n",
    "    print(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note, that the chunksize keyword applies to the **source** rows. So if you\n",
    "are doing a query, then the chunksize will subdivide the total rows in the table\n",
    "and the query applied, returning an iterator on potentially unequal sized chunks.\n",
    "\n",
    "Here is a recipe for generating a query and using it to create equal sized return\n",
    "chunks."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Advanced queries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Select a single column\n",
    "\n",
    "To retrieve a single indexable or data column, use the\n",
    "method `select_column`. This will, for example, enable you to get the index\n",
    "very quickly. These return a `Series` of the result, indexed by the row number.\n",
    "These do not currently accept the `where` selector.\n",
    "\n",
    "\n",
    "<a id='io-hdf5-selecting-coordinates'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Selecting coordinates\n",
    "\n",
    "Sometimes you want to get the coordinates (a.k.a the index locations) of your query. This returns an\n",
    "`Int64Index` of the resulting locations. These coordinates can also be passed to subsequent\n",
    "`where` operations.\n",
    "\n",
    "\n",
    "<a id='io-hdf5-where-mask'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Selecting using a where mask\n",
    "\n",
    "Sometime your query can involve creating a list of rows to select. Usually this `mask` would\n",
    "be a resulting `index` from an indexing operation. This example selects the months of\n",
    "a datetimeindex which are 5."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Storer object\n",
    "\n",
    "If you want to inspect the stored object, retrieve via\n",
    "`get_storer`. You could use this programmatically to say get the number\n",
    "of rows in an object."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Multiple table queries\n",
    "\n",
    "The methods `append_to_multiple` and\n",
    "`select_as_multiple` can perform appending/selecting from\n",
    "multiple tables at once. The idea is to have one table (call it the\n",
    "selector table) that you index most/all of the columns, and perform your\n",
    "queries. The other table(s) are data tables with an index matching the\n",
    "selector table’s index. You can then perform a very fast query\n",
    "on the selector table, yet get lots of data back. This method is similar to\n",
    "having a very wide table, but enables more efficient queries.\n",
    "\n",
    "The `append_to_multiple` method splits a given single DataFrame\n",
    "into multiple tables according to `d`, a dictionary that maps the\n",
    "table names to a list of ‘columns’ you want in that table. If None\n",
    "is used in place of a list, that table will have the remaining\n",
    "unspecified columns of the given DataFrame. The argument `selector`\n",
    "defines which table is the selector table (which you can make queries from).\n",
    "The argument `dropna` will drop rows from the input `DataFrame` to ensure\n",
    "tables are synchronized.  This means that if a row for one of the tables\n",
    "being written to is entirely `np.NaN`, that row will be dropped from all tables.\n",
    "\n",
    "If `dropna` is False, **THE USER IS RESPONSIBLE FOR SYNCHRONIZING THE TABLES**.\n",
    "Remember that entirely `np.Nan` rows are not written to the HDFStore, so if\n",
    "you choose to call `dropna=False`, some tables may have more rows than others,\n",
    "and therefore `select_as_multiple` may not work or it may return unexpected\n",
    "results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Delete from a table\n",
    "\n",
    "You can delete from a table selectively by specifying a `where`. In\n",
    "deleting rows, it is important to understand the `PyTables` deletes\n",
    "rows by erasing the rows, then **moving** the following data. Thus\n",
    "deleting can potentially be a very expensive operation depending on the\n",
    "orientation of your data. To get optimal performance, it’s\n",
    "worthwhile to have the dimension you are deleting be the first of the\n",
    "`indexables`.\n",
    "\n",
    "Data is ordered (on the disk) in terms of the `indexables`. Here’s a\n",
    "simple use case. You store panel-type data, with dates in the\n",
    "`major_axis` and ids in the `minor_axis`. The data is then\n",
    "interleaved like this:\n",
    "\n",
    "- \n",
    "  <dl style='margin: 20px 0;'>\n",
    "  <dt>date_1</dt>\n",
    "  <dd>\n",
    "  - id_1  \n",
    "  - id_2  \n",
    "  - .  \n",
    "  - id_n  \n",
    "  </dd>\n",
    "  \n",
    "  </dl>\n",
    "  \n",
    "- \n",
    "  <dl style='margin: 20px 0;'>\n",
    "  <dt>date_2</dt>\n",
    "  <dd>\n",
    "  - id_1  \n",
    "  - .  \n",
    "  - id_n  \n",
    "  </dd>\n",
    "  \n",
    "  </dl>\n",
    "  \n",
    "\n",
    "\n",
    "It should be clear that a delete operation on the `major_axis` will be\n",
    "fairly quick, as one chunk is removed, then the following data moved. On\n",
    "the other hand a delete operation on the `minor_axis` will be very\n",
    "expensive. In this case it would almost certainly be faster to rewrite\n",
    "the table using a `where` that selects all but the missing data.\n",
    "\n",
    "Please note that HDF5 **DOES NOT RECLAIM SPACE** in the h5 files\n",
    "automatically. Thus, repeatedly deleting (or removing nodes) and adding\n",
    "again, **WILL TEND TO INCREASE THE FILE SIZE**.\n",
    "\n",
    "To *repack and clean* the file, use [ptrepack](#io-hdf5-ptrepack).\n",
    "\n",
    "\n",
    "<a id='io-hdf5-notes'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notes & caveats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compression\n",
    "\n",
    "`PyTables` allows the stored data to be compressed. This applies to\n",
    "all kinds of stores, not just tables. Two parameters are used to\n",
    "control compression: `complevel` and `complib`.\n",
    "\n",
    "\n",
    "<dl style='margin: 20px 0;'>\n",
    "<dt>`complevel` specifies if and how hard data is to be compressed.</dt>\n",
    "<dd>\n",
    "`complevel=0` and `complevel=None` disables\n",
    "compression and `0<complevel<10` enables compression.\n",
    "\n",
    "</dd>\n",
    "<dt>`complib` specifies which compression library to use. If nothing is</dt>\n",
    "<dd>\n",
    "specified the default library `zlib` is used. A\n",
    "compression library usually optimizes for either good\n",
    "compression rates or speed and the results will depend on\n",
    "the type of data. Which type of\n",
    "compression to choose depends on your specific needs and\n",
    "data. The list of supported compression libraries:\n",
    "\n",
    "> - [zlib](https://zlib.net/): The default compression library. A classic in terms of compression, achieves good compression rates but is somewhat slow.  \n",
    "- [lzo](https://www.oberhumer.com/opensource/lzo/): Fast compression and decompression.  \n",
    "- [bzip2](http://bzip.org/): Good compression rates.  \n",
    "- [blosc](http://www.blosc.org/): Fast compression and decompression.  \n",
    "\n",
    "\n",
    "New in version 0.20.2: Support for alternative blosc compressors:\n",
    "\n",
    "- [blosc:blosclz](http://www.blosc.org/) This is the\n",
    "  default compressor for `blosc`  \n",
    "- [blosc:lz4](https://fastcompression.blogspot.dk/p/lz4.html):\n",
    "  A compact, very popular and fast compressor.  \n",
    "- [blosc:lz4hc](https://fastcompression.blogspot.dk/p/lz4.html):\n",
    "  A tweaked version of LZ4, produces better\n",
    "  compression ratios at the expense of speed.  \n",
    "- [blosc:snappy](https://google.github.io/snappy/):\n",
    "  A popular compressor used in many places.  \n",
    "- [blosc:zlib](https://zlib.net/): A classic;\n",
    "  somewhat slower than the previous ones, but\n",
    "  achieving better compression ratios.  \n",
    "- [blosc:zstd](https://facebook.github.io/zstd/): An\n",
    "  extremely well balanced codec; it provides the best\n",
    "  compression ratios among the others above, and at\n",
    "  reasonably fast speed.  \n",
    "\n",
    "\n",
    "If `complib` is defined as something other than the\n",
    "listed libraries a `ValueError` exception is issued.\n",
    "\n",
    "\n",
    "</dd>\n",
    "\n",
    "</dl>\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">If the library specified with the `complib` option is missing on your platform,\n",
    "compression defaults to `zlib` without further ado.\n",
    "\n",
    "Enable compression for all objects within the file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "store_compressed = pd.HDFStore('store_compressed.h5', complevel=9,\n",
    "                               complib='blosc:blosclz')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or on-the-fly compression (this only applies to tables) in stores where compression is not enabled:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "store.append('df', df, complib='zlib', complevel=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<a id='io-hdf5-ptrepack'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ptrepack\n",
    "\n",
    "`PyTables` offers better write performance when tables are compressed after\n",
    "they are written, as opposed to turning on compression at the very\n",
    "beginning. You can use the supplied `PyTables` utility\n",
    "`ptrepack`. In addition, `ptrepack` can change compression levels\n",
    "after the fact."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hide-output": false
   },
   "source": [
    "```console\n",
    "ptrepack --chunkshape=auto --propindexes --complevel=9 --complib=blosc in.h5 out.h5\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Furthermore `ptrepack in.h5 out.h5` will *repack* the file to allow\n",
    "you to reuse previously deleted space. Alternatively, one can simply\n",
    "remove the file and write again, or use the `copy` method.\n",
    "\n",
    "\n",
    "<a id='io-hdf5-caveats'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Caveats\n",
    "\n",
    "`HDFStore` is **not-threadsafe for writing**. The underlying\n",
    "`PyTables` only supports concurrent reads (via threading or\n",
    "processes). If you need reading and writing *at the same time*, you\n",
    "need to serialize these operations in a single thread in a single\n",
    "process. You will corrupt your data otherwise. See the ([GH2397](https://github.com/pandas-dev/pandas/issues/2397)) for more information.\n",
    "\n",
    "- If you use locks to manage write access between multiple processes, you\n",
    "  may want to use [`fsync()`](https://docs.python.org/3/library/os.html#os.fsync) before releasing write locks. For\n",
    "  convenience you can use `store.flush(fsync=True)` to do this for you.  \n",
    "- Once a `table` is created columns (DataFrame)\n",
    "  are fixed; only exactly the same columns can be appended  \n",
    "- Be aware that timezones (e.g., `pytz.timezone('US/Eastern')`)\n",
    "  are not necessarily equal across timezone versions.  So if data is\n",
    "  localized to a specific timezone in the HDFStore using one version\n",
    "  of a timezone library and that data is updated with another version, the data\n",
    "  will be converted to UTC since these timezones are not considered\n",
    "  equal.  Either use the same version of timezone library or use `tz_convert` with\n",
    "  the updated timezone definition.  \n",
    "\n",
    "\n",
    "`PyTables` will show a `NaturalNameWarning` if a column name\n",
    "cannot be used as an attribute selector.\n",
    "*Natural* identifiers contain only letters, numbers, and underscores,\n",
    "and may not begin with a number.\n",
    "Other identifiers cannot be used in a `where` clause\n",
    "and are generally a bad idea.\n",
    "\n",
    "\n",
    "<a id='io-hdf5-data-types'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DataTypes\n",
    "\n",
    "`HDFStore` will map an object dtype to the `PyTables` underlying\n",
    "dtype. This means the following types are known to work:\n",
    "\n",
    "``````````````````|Type|Represents missing values|\n",
    "|:----------------------------------------------------:|:-----------------------:|\n",
    "|floating : float64, float32, float16|np.nan|\n",
    "|integer : int64, int32, int8, uint64,uint32, uint8||\n",
    "|boolean||\n",
    "|datetime64[ns]|NaT|\n",
    "|timedelta64[ns]|NaT|\n",
    "|categorical : see the section below||\n",
    "|object : strings|np.nan|\n",
    "`unicode` columns are not supported, and **WILL FAIL**.\n",
    "\n",
    "\n",
    "<a id='io-hdf5-categorical'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Categorical data\n",
    "\n",
    "You can write data that contains `category` dtypes to a `HDFStore`.\n",
    "Queries work the same as if it was an object array. However, the `category` dtyped data is\n",
    "stored in a more efficient manner."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### String columns\n",
    "\n",
    "**min_itemsize**\n",
    "\n",
    "The underlying implementation of `HDFStore` uses a fixed column width (itemsize) for string columns.\n",
    "A string column itemsize is calculated as the maximum of the\n",
    "length of data (for that column) that is passed to the `HDFStore`, **in the first append**. Subsequent appends,\n",
    "may introduce a string for a column **larger** than the column can hold, an Exception will be raised (otherwise you\n",
    "could have a silent truncation of these columns, leading to loss of information). In the future we may relax this and\n",
    "allow a user-specified truncation to occur.\n",
    "\n",
    "Pass `min_itemsize` on the first table creation to a-priori specify the minimum length of a particular string column.\n",
    "`min_itemsize` can be an integer, or a dict mapping a column name to an integer. You can pass `values` as a key to\n",
    "allow all *indexables* or *data_columns* to have this min_itemsize.\n",
    "\n",
    "Passing a `min_itemsize` dict will cause all passed columns to be created as *data_columns* automatically.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">If you are not passing any `data_columns`, then the `min_itemsize` will be the maximum of the length of any string passed\n",
    "\n",
    "**nan_rep**\n",
    "\n",
    "String columns will serialize a `np.nan` (a missing value) with the `nan_rep` string representation. This defaults to the string value `nan`.\n",
    "You could inadvertently turn an actual `nan` value into a missing value.\n",
    "\n",
    "\n",
    "<a id='io-external-compatibility'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### External compatibility\n",
    "\n",
    "`HDFStore` writes `table` format objects in specific formats suitable for\n",
    "producing loss-less round trips to pandas objects. For external\n",
    "compatibility, `HDFStore` can read native `PyTables` format\n",
    "tables.\n",
    "\n",
    "It is possible to write an `HDFStore` object that can easily be imported into `R` using the\n",
    "`rhdf5` library ([Package website](https://www.bioconductor.org/packages/release/bioc/html/rhdf5.html)). Create a table format store like this:\n",
    "\n",
    "In R this file can be read into a `data.frame` object using the `rhdf5`\n",
    "library. The following example function reads the corresponding column names\n",
    "and data values from the values and assembles them into a `data.frame`:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hide-output": false
   },
   "source": [
    "```R\n",
    "# Load values and column names for all datasets from corresponding nodes and\n",
    "# insert them into one data.frame object.\n",
    "\n",
    "library(rhdf5)\n",
    "\n",
    "loadhdf5data <- function(h5File) {\n",
    "\n",
    "listing <- h5ls(h5File)\n",
    "# Find all data nodes, values are stored in *_values and corresponding column\n",
    "# titles in *_items\n",
    "data_nodes <- grep(\"_values\", listing$name)\n",
    "name_nodes <- grep(\"_items\", listing$name)\n",
    "data_paths = paste(listing$group[data_nodes], listing$name[data_nodes], sep = \"/\")\n",
    "name_paths = paste(listing$group[name_nodes], listing$name[name_nodes], sep = \"/\")\n",
    "columns = list()\n",
    "for (idx in seq(data_paths)) {\n",
    "  # NOTE: matrices returned by h5read have to be transposed to obtain\n",
    "  # required Fortran order!\n",
    "  data <- data.frame(t(h5read(h5File, data_paths[idx])))\n",
    "  names <- t(h5read(h5File, name_paths[idx]))\n",
    "  entry <- data.frame(data)\n",
    "  colnames(entry) <- names\n",
    "  columns <- append(columns, entry)\n",
    "}\n",
    "\n",
    "data <- data.frame(columns)\n",
    "\n",
    "return(data)\n",
    "}\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now you can import the `DataFrame` into R:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hide-output": false
   },
   "source": [
    "```R\n",
    "> data = loadhdf5data(\"transfer.hdf5\")\n",
    "> head(data)\n",
    "         first    second class\n",
    "1 0.4170220047 0.3266449     0\n",
    "2 0.7203244934 0.5270581     0\n",
    "3 0.0001143748 0.8859421     1\n",
    "4 0.3023325726 0.3572698     1\n",
    "5 0.1467558908 0.9085352     1\n",
    "6 0.0923385948 0.6233601     1\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ">**Note**\n",
    ">\n",
    ">The R function lists the entire HDF5 file’s contents and assembles the\n",
    "`data.frame` object from all matching nodes, so use this only as a\n",
    "starting point if you have stored multiple `DataFrame` objects to a\n",
    "single HDF5 file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Performance\n",
    "\n",
    "- `tables` format come with a writing performance penalty as compared to\n",
    "  `fixed` stores. The benefit is the ability to append/delete and\n",
    "  query (potentially very large amounts of data).  Write times are\n",
    "  generally longer as compared with regular stores. Query times can\n",
    "  be quite fast, especially on an indexed axis.  \n",
    "- You can pass `chunksize=<int>` to `append`, specifying the\n",
    "  write chunksize (default is 50000). This will significantly lower\n",
    "  your memory usage on writing.  \n",
    "- You can pass `expectedrows=<int>` to the first `append`,\n",
    "  to set the TOTAL number of expected rows that `PyTables` will\n",
    "  expected. This will optimize read/write performance.  \n",
    "- Duplicate rows can be written to tables, but are filtered out in\n",
    "  selection (with the last items being selected; thus a table is\n",
    "  unique on major, minor pairs)  \n",
    "- A `PerformanceWarning` will be raised if you are attempting to\n",
    "  store types that will be pickled by PyTables (rather than stored as\n",
    "  endemic types). See\n",
    "  [Here](https://stackoverflow.com/questions/14355151/how-to-make-pandas-hdfstore-put-operation-faster/14370190#14370190)\n",
    "  for more information and some solutions.  \n",
    "\n",
    "\n",
    "\n",
    "<a id='io-feather'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Feather\n",
    "\n",
    "New in version 0.20.0.\n",
    "\n",
    "Feather provides binary columnar serialization for data frames. It is designed to make reading and writing data\n",
    "frames efficient, and to make sharing data across data analysis languages easy.\n",
    "\n",
    "Feather is designed to faithfully serialize and de-serialize DataFrames, supporting all of the pandas\n",
    "dtypes, including extension dtypes such as categorical and datetime with tz.\n",
    "\n",
    "Several caveats.\n",
    "\n",
    "- This is a newer library, and the format, though stable, is not guaranteed to be backward compatible\n",
    "  to the earlier versions.  \n",
    "- The format will NOT write an `Index`, or `MultiIndex` for the\n",
    "  `DataFrame` and will raise an error if a non-default one is provided. You\n",
    "  can `.reset_index()` to store the index or `.reset_index(drop=True)` to\n",
    "  ignore it.  \n",
    "- Duplicate column names and non-string columns names are not supported  \n",
    "- Non supported types include `Period` and actual Python object types. These will raise a helpful error message\n",
    "  on an attempt at serialization.  \n",
    "\n",
    "\n",
    "See the [Full Documentation](https://github.com/wesm/feather).\n",
    "\n",
    "Write to a feather file.\n",
    "\n",
    "Read from a feather file.\n",
    "\n",
    "\n",
    "<a id='io-parquet'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parquet\n",
    "\n",
    "New in version 0.21.0.\n",
    "\n",
    "[Apache Parquet](https://parquet.apache.org/) provides a partitioned binary columnar serialization for data frames. It is designed to\n",
    "make reading and writing data frames efficient, and to make sharing data across data analysis\n",
    "languages easy. Parquet can use a variety of compression techniques to shrink the file size as much as possible\n",
    "while still maintaining good read performance.\n",
    "\n",
    "Parquet is designed to faithfully serialize and de-serialize `DataFrame` s, supporting all of the pandas\n",
    "dtypes, including extension dtypes such as datetime with tz.\n",
    "\n",
    "Several caveats.\n",
    "\n",
    "- Duplicate column names and non-string columns names are not supported.  \n",
    "- The `pyarrow` engine always writes the index to the output, but `fastparquet` only writes non-default\n",
    "  indexes. This extra column can cause problems for non-Pandas consumers that are not expecting it. You can\n",
    "  force including or omitting indexes with the `index` argument, regardless of the underlying engine.  \n",
    "- Index level names, if specified, must be strings.  \n",
    "- Categorical dtypes can be serialized to parquet, but will de-serialize as `object` dtype.  \n",
    "- Non supported types include `Period` and actual Python object types. These will raise a helpful error message\n",
    "  on an attempt at serialization.  \n",
    "\n",
    "\n",
    "You can specify an `engine` to direct the serialization. This can be one of `pyarrow`, or `fastparquet`, or `auto`.\n",
    "If the engine is NOT specified, then the `pd.options.io.parquet.engine` option is checked; if this is also `auto`,\n",
    "then `pyarrow` is tried, and falling back to `fastparquet`.\n",
    "\n",
    "See the documentation for [pyarrow](https://arrow.apache.org/docs/python/) and [fastparquet](https://fastparquet.readthedocs.io/en/latest/).\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">These engines are very similar and should read/write nearly identical parquet format files.\n",
    "Currently `pyarrow` does not support timedelta data, `fastparquet>=0.1.4` supports timezone aware datetimes.\n",
    "These libraries differ by having different underlying dependencies (`fastparquet` by using `numba`, while `pyarrow` uses a c-library).\n",
    "\n",
    "Write to a parquet file.\n",
    "\n",
    "Read from a parquet file.\n",
    "\n",
    "Read only certain columns of a parquet file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Handling indexes\n",
    "\n",
    "Serializing a `DataFrame` to parquet may include the implicit index as one or\n",
    "more columns in the output file. Thus, this code:\n",
    "\n",
    "creates a parquet file with *three* columns if you use `pyarrow` for serialization:\n",
    "`a`, `b`, and `__index_level_0__`. If you’re using `fastparquet`, the\n",
    "index [may or may not](https://fastparquet.readthedocs.io/en/latest/api.html#fastparquet.write)\n",
    "be written to the file.\n",
    "\n",
    "This unexpected extra column causes some databases like Amazon Redshift to reject\n",
    "the file, because that column doesn’t exist in the target table.\n",
    "\n",
    "If you want to omit a dataframe’s indexes when writing, pass `index=False` to\n",
    "`to_parquet()`:\n",
    "\n",
    "This creates a parquet file with just the two expected columns, `a` and `b`.\n",
    "If your `DataFrame` has a custom index, you won’t get it back when you load\n",
    "this file into a `DataFrame`.\n",
    "\n",
    "Passing `index=True` will *always* write the index, even if that’s not the\n",
    "underlying engine’s default behavior."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Partitioning Parquet files\n",
    "\n",
    "New in version 0.24.0.\n",
    "\n",
    "Parquet supports partitioning of data based on the values of one or more columns.\n",
    "\n",
    "The fname specifies the parent directory to which data will be saved.\n",
    "The partition_cols are the column names by which the dataset will be partitioned.\n",
    "Columns are partitioned in the order they are given. The partition splits are\n",
    "determined by the unique values in the partition columns.\n",
    "The above example creates a partitioned dataset that may look like:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hide-output": false
   },
   "source": [
    "```text\n",
    "test\n",
    "├── a=0\n",
    "│   ├── 0bac803e32dc42ae83fddfd029cbdebc.parquet\n",
    "│   └──  ...\n",
    "└── a=1\n",
    "    ├── e6ab24a4f45147b49b54a662f0c412a3.parquet\n",
    "    └── ...\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<a id='io-sql'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SQL queries\n",
    "\n",
    "The `pandas.io.sql` module provides a collection of query wrappers to both\n",
    "facilitate data retrieval and to reduce dependency on DB-specific API. Database abstraction\n",
    "is provided by SQLAlchemy if installed. In addition you will need a driver library for\n",
    "your database. Examples of such drivers are [psycopg2](http://initd.org/psycopg/)\n",
    "for PostgreSQL or [pymysql](https://github.com/PyMySQL/PyMySQL) for MySQL.\n",
    "For [SQLite](https://docs.python.org/3/library/sqlite3.html) this is\n",
    "included in Python’s standard library by default.\n",
    "You can find an overview of supported drivers for each SQL dialect in the\n",
    "[SQLAlchemy docs](https://docs.sqlalchemy.org/en/latest/dialects/index.html).\n",
    "\n",
    "If SQLAlchemy is not installed, a fallback is only provided for sqlite (and\n",
    "for mysql for backwards compatibility, but this is deprecated and will be\n",
    "removed in a future version).\n",
    "This mode requires a Python database adapter which respect the [Python\n",
    "DB-API](https://www.python.org/dev/peps/pep-0249/).\n",
    "\n",
    "See also some [cookbook examples](cookbook.ipynb#cookbook-sql) for some advanced strategies.\n",
    "\n",
    "The key functions are:\n",
    "\n",
    "[``](reference/api/pandas.read_sql_table.ipynb#pandas.read_sql_table)[``](reference/api/pandas.read_sql_query.ipynb#pandas.read_sql_query)[``](reference/api/pandas.read_sql.ipynb#pandas.read_sql)[``](reference/api/pandas.DataFrame.to_sql.ipynb#pandas.DataFrame.to_sql)|read_sql_table(table_name, con[, schema, …])|Read SQL database table into a DataFrame.|\n",
    "|read_sql_query(sql, con[, index_col, …])|Read SQL query into a DataFrame.|\n",
    "|read_sql(sql, con[, index_col, …])|Read SQL query or database table into a DataFrame.|\n",
    "|DataFrame.to_sql(self, name, con[, schema, …])|Write records stored in a DataFrame to a SQL database.|\n",
    "- [pandas.read_sql_table](reference/api/pandas.read_sql_table.ipynb)\n",
    "- [pandas.read_sql_query](reference/api/pandas.read_sql_query.ipynb)\n",
    "- [pandas.read_sql](reference/api/pandas.read_sql.ipynb)\n",
    "- [pandas.DataFrame.to_sql](reference/api/pandas.DataFrame.to_sql.ipynb)\n",
    "\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">The function [`read_sql()`](reference/api/pandas.read_sql.ipynb#pandas.read_sql) is a convenience wrapper around\n",
    "[`read_sql_table()`](reference/api/pandas.read_sql_table.ipynb#pandas.read_sql_table) and [`read_sql_query()`](reference/api/pandas.read_sql_query.ipynb#pandas.read_sql_query) (and for\n",
    "backward compatibility) and will delegate to specific function depending on\n",
    "the provided input (database table name or sql query).\n",
    "Table names do not need to be quoted if they have special characters.\n",
    "\n",
    "In the following example, we use the [SQlite](https://www.sqlite.org/) SQL database\n",
    "engine. You can use a temporary SQLite database where data are stored in\n",
    "“memory”.\n",
    "\n",
    "To connect with SQLAlchemy you use the `create_engine()` function to create an engine\n",
    "object from database URI. You only need to create the engine once per database you are\n",
    "connecting to.\n",
    "For more information on `create_engine()` and the URI formatting, see the examples\n",
    "below and the SQLAlchemy [documentation](https://docs.sqlalchemy.org/en/latest/core/engines.html)\n",
    "\n",
    "If you want to manage your own connections you can pass one of those instead:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "with engine.connect() as conn, conn.begin():\n",
    "    data = pd.read_sql_table('data', conn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing DataFrames\n",
    "\n",
    "Assuming the following data is in a `DataFrame` `data`, we can insert it into\n",
    "the database using [`to_sql()`](reference/api/pandas.DataFrame.to_sql.ipynb#pandas.DataFrame.to_sql).\n",
    "\n",
    "|id|Date|Col_1|Col_2|Col_3|\n",
    "|:---:|:----------:|:-----:|:-----:|:-----:|\n",
    "|26|2012-10-18|X|25.7|True|\n",
    "|42|2012-10-19|Y|-12.4|False|\n",
    "|63|2012-10-20|Z|5.73|True|\n",
    "With some databases, writing large DataFrames can result in errors due to\n",
    "packet size limitations being exceeded. This can be avoided by setting the\n",
    "`chunksize` parameter when calling `to_sql`.  For example, the following\n",
    "writes `data` to the database in batches of 1000 rows at a time:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### SQL data types\n",
    "\n",
    "[`to_sql()`](reference/api/pandas.DataFrame.to_sql.ipynb#pandas.DataFrame.to_sql) will try to map your data to an appropriate\n",
    "SQL data type based on the dtype of the data. When you have columns of dtype\n",
    "`object`, pandas will try to infer the data type.\n",
    "\n",
    "You can always override the default type by specifying the desired SQL type of\n",
    "any of the columns by using the `dtype` argument. This argument needs a\n",
    "dictionary mapping column names to SQLAlchemy types (or strings for the sqlite3\n",
    "fallback mode).\n",
    "For example, specifying to use the sqlalchemy `String` type instead of the\n",
    "default `Text` type for string columns:\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">Due to the limited support for timedelta’s in the different database\n",
    "flavors, columns with type `timedelta64` will be written as integer\n",
    "values as nanoseconds to the database and a warning will be raised.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">Columns of `category` dtype will be converted to the dense representation\n",
    "as you would get with `np.asarray(categorical)` (e.g. for string categories\n",
    "this gives an array of strings).\n",
    "Because of this, reading the database table back in does **not** generate\n",
    "a categorical.\n",
    "\n",
    "\n",
    "<a id='io-sql-datetime-data'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Datetime data types\n",
    "\n",
    "Using SQLAlchemy, [`to_sql()`](reference/api/pandas.DataFrame.to_sql.ipynb#pandas.DataFrame.to_sql) is capable of writing\n",
    "datetime data that is timezone naive or timezone aware. However, the resulting\n",
    "data stored in the database ultimately depends on the supported data type\n",
    "for datetime data of the database system being used.\n",
    "\n",
    "The following table lists supported data types for datetime data for some\n",
    "common databases. Other database dialects may have different data types for\n",
    "datetime data.\n",
    "\n",
    "``````````|Database|SQL Datetime Types|Timezone Support|\n",
    "|:---------:|:-------------------------------------------:|:-----------------:|\n",
    "|SQLite|TEXT|No|\n",
    "|MySQL|TIMESTAMP or DATETIME|No|\n",
    "|PostgreSQL|TIMESTAMP or TIMESTAMP WITH TIME ZONE|Yes|\n",
    "When writing timezone aware data to databases that do not support timezones,\n",
    "the data will be written as timezone naive timestamps that are in local time\n",
    "with respect to the timezone.\n",
    "\n",
    "[`read_sql_table()`](reference/api/pandas.read_sql_table.ipynb#pandas.read_sql_table) is also capable of reading datetime data that is\n",
    "timezone aware or naive. When reading `TIMESTAMP WITH TIME ZONE` types, pandas\n",
    "will convert the data to UTC.\n",
    "\n",
    "\n",
    "<a id='io-sql-method'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Insertion method\n",
    "\n",
    "New in version 0.24.0.\n",
    "\n",
    "The parameter `method` controls the SQL insertion clause used.\n",
    "Possible values are:\n",
    "\n",
    "- `None`: Uses standard SQL `INSERT` clause (one per row).  \n",
    "- `'multi'`: Pass multiple values in a single `INSERT` clause.\n",
    "  It uses a *special* SQL syntax not supported by all backends.\n",
    "  This usually provides better performance for analytic databases\n",
    "  like *Presto* and *Redshift*, but has worse performance for\n",
    "  traditional SQL backend if the table contains many columns.\n",
    "  For more information check the SQLAlchemy [documention](http://docs.sqlalchemy.org/en/latest/core/dml.html#sqlalchemy.sql.expression.Insert.values.params.*args).  \n",
    "- callable with signature `(pd_table, conn, keys, data_iter)`:\n",
    "  This can be used to implement a more performant insertion method based on\n",
    "  specific backend dialect features.  \n",
    "\n",
    "\n",
    "Example of a callable using PostgreSQL [COPY clause](https://www.postgresql.org/docs/current/static/sql-copy.html):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "# Alternative to_sql() *method* for DBs that support COPY FROM\n",
    "import csv\n",
    "from io import StringIO\n",
    "\n",
    "def psql_insert_copy(table, conn, keys, data_iter):\n",
    "    # gets a DBAPI connection that can provide a cursor\n",
    "    dbapi_conn = conn.connection\n",
    "    with dbapi_conn.cursor() as cur:\n",
    "        s_buf = StringIO()\n",
    "        writer = csv.writer(s_buf)\n",
    "        writer.writerows(data_iter)\n",
    "        s_buf.seek(0)\n",
    "\n",
    "        columns = ', '.join('\"{}\"'.format(k) for k in keys)\n",
    "        if table.schema:\n",
    "            table_name = '{}.{}'.format(table.schema, table.name)\n",
    "        else:\n",
    "            table_name = table.name\n",
    "\n",
    "        sql = 'COPY {} ({}) FROM STDIN WITH CSV'.format(\n",
    "            table_name, columns)\n",
    "        cur.copy_expert(sql=sql, file=s_buf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading tables\n",
    "\n",
    "[`read_sql_table()`](reference/api/pandas.read_sql_table.ipynb#pandas.read_sql_table) will read a database table given the\n",
    "table name and optionally a subset of columns to read.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">In order to use [`read_sql_table()`](reference/api/pandas.read_sql_table.ipynb#pandas.read_sql_table), you **must** have the\n",
    "SQLAlchemy optional dependency installed.\n",
    "\n",
    "You can also specify the name of the column as the `DataFrame` index,\n",
    "and specify a subset of columns to be read.\n",
    "\n",
    "And you can explicitly force columns to be parsed as dates:\n",
    "\n",
    "If needed you can explicitly specify a format string, or a dict of arguments\n",
    "to pass to `pandas.to_datetime()`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "pd.read_sql_table('data', engine, parse_dates={'Date': '%Y-%m-%d'})\n",
    "pd.read_sql_table('data', engine,\n",
    "                  parse_dates={'Date': {'format': '%Y-%m-%d %H:%M:%S'}})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can check if a table exists using `has_table()`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Schema support\n",
    "\n",
    "Reading from and writing to different schema’s is supported through the `schema`\n",
    "keyword in the [`read_sql_table()`](reference/api/pandas.read_sql_table.ipynb#pandas.read_sql_table) and [`to_sql()`](reference/api/pandas.DataFrame.to_sql.ipynb#pandas.DataFrame.to_sql)\n",
    "functions. Note however that this depends on the database flavor (sqlite does not\n",
    "have schema’s). For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "df.to_sql('table', engine, schema='other_schema')\n",
    "pd.read_sql_table('table', engine, schema='other_schema')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Querying\n",
    "\n",
    "You can query using raw SQL in the [`read_sql_query()`](reference/api/pandas.read_sql_query.ipynb#pandas.read_sql_query) function.\n",
    "In this case you must use the SQL variant appropriate for your database.\n",
    "When using SQLAlchemy, you can also pass SQLAlchemy Expression language constructs,\n",
    "which are database-agnostic.\n",
    "\n",
    "Of course, you can specify a more “complex” query.\n",
    "\n",
    "The [`read_sql_query()`](reference/api/pandas.read_sql_query.ipynb#pandas.read_sql_query) function supports a `chunksize` argument.\n",
    "Specifying this will return an iterator through chunks of the query result:\n",
    "\n",
    "You can also run a plain query without creating a `DataFrame` with\n",
    "`execute()`. This is useful for queries that don’t return values,\n",
    "such as INSERT. This is functionally equivalent to calling `execute` on the\n",
    "SQLAlchemy engine or db connection object. Again, you must use the SQL syntax\n",
    "variant appropriate for your database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "from pandas.io import sql\n",
    "sql.execute('SELECT * FROM table_name', engine)\n",
    "sql.execute('INSERT INTO table_name VALUES(?, ?, ?)', engine,\n",
    "            params=[('id', 1, 12.2, True)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Engine connection examples\n",
    "\n",
    "To connect with SQLAlchemy you use the `create_engine()` function to create an engine\n",
    "object from database URI. You only need to create the engine once per database you are\n",
    "connecting to."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "from sqlalchemy import create_engine\n",
    "\n",
    "engine = create_engine('postgresql://scott:tiger@localhost:5432/mydatabase')\n",
    "\n",
    "engine = create_engine('mysql+mysqldb://scott:tiger@localhost/foo')\n",
    "\n",
    "engine = create_engine('oracle://scott:tiger@127.0.0.1:1521/sidname')\n",
    "\n",
    "engine = create_engine('mssql+pyodbc://mydsn')\n",
    "\n",
    "# sqlite://<nohostname>/<path>\n",
    "# where <path> is relative:\n",
    "engine = create_engine('sqlite:///foo.db')\n",
    "\n",
    "# or absolute, starting with a slash:\n",
    "engine = create_engine('sqlite:////absolute/path/to/foo.db')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For more information see the examples the SQLAlchemy [documentation](https://docs.sqlalchemy.org/en/latest/core/engines.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Advanced SQLAlchemy queries\n",
    "\n",
    "You can use SQLAlchemy constructs to describe your query.\n",
    "\n",
    "Use `sqlalchemy.text()` to specify query parameters in a backend-neutral way\n",
    "\n",
    "If you have an SQLAlchemy description of your database you can express where conditions using SQLAlchemy expressions\n",
    "\n",
    "You can combine SQLAlchemy expressions with parameters passed to [`read_sql()`](reference/api/pandas.read_sql.ipynb#pandas.read_sql) using `sqlalchemy.bindparam()`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sqlite fallback\n",
    "\n",
    "The use of sqlite is supported without using SQLAlchemy.\n",
    "This mode requires a Python database adapter which respect the [Python\n",
    "DB-API](https://www.python.org/dev/peps/pep-0249/).\n",
    "\n",
    "You can create connections like so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "import sqlite3\n",
    "con = sqlite3.connect(':memory:')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And then issue the following queries:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "data.to_sql('data', con)\n",
    "pd.read_sql_query(\"SELECT * FROM data\", con)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<a id='io-bigquery'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Google BigQuery\n",
    "\n",
    "Starting in 0.20.0, pandas has split off Google BigQuery support into the\n",
    "separate package `pandas-gbq`. You can `pip install pandas-gbq` to get it.\n",
    "\n",
    "The `pandas-gbq` package provides functionality to read/write from Google BigQuery.\n",
    "\n",
    "pandas integrates with this external package. if `pandas-gbq` is installed, you can\n",
    "use the pandas methods `pd.read_gbq` and `DataFrame.to_gbq`, which will call the\n",
    "respective functions from `pandas-gbq`.\n",
    "\n",
    "Full documentation can be found [here](https://pandas-gbq.readthedocs.io/).\n",
    "\n",
    "\n",
    "<a id='io-stata'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stata format\n",
    "\n",
    "\n",
    "<a id='io-stata-writer'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing to stata format\n",
    "\n",
    "The method `to_stata()` will write a DataFrame\n",
    "into a .dta file. The format version of this file is always 115 (Stata 12).\n",
    "\n",
    "*Stata* data files have limited data type support; only strings with\n",
    "244 or fewer characters, `int8`, `int16`, `int32`, `float32`\n",
    "and `float64` can be stored in `.dta` files.  Additionally,\n",
    "*Stata* reserves certain values to represent missing data. Exporting a\n",
    "non-missing value that is outside of the permitted range in Stata for\n",
    "a particular data type will retype the variable to the next larger\n",
    "size.  For example, `int8` values are restricted to lie between -127\n",
    "and 100 in Stata, and so variables with values above 100 will trigger\n",
    "a conversion to `int16`. `nan` values in floating points data\n",
    "types are stored as the basic missing data type (`.` in *Stata*).\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">It is not possible to export missing data values for integer data types.\n",
    "\n",
    "The *Stata* writer gracefully handles other data types including `int64`,\n",
    "`bool`, `uint8`, `uint16`, `uint32` by casting to\n",
    "the smallest supported type that can represent the data.  For example, data\n",
    "with a type of `uint8` will be cast to `int8` if all values are less than\n",
    "100 (the upper bound for non-missing `int8` data in *Stata*), or, if values are\n",
    "outside of this range, the variable is cast to `int16`.\n",
    "\n",
    "Conversion from `int64` to `float64` may result in a loss of precision\n",
    "if `int64` values are larger than 2**53.\n",
    "\n",
    "`StataWriter` and\n",
    "`to_stata()` only support fixed width\n",
    "strings containing up to 244 characters, a limitation imposed by the version\n",
    "115 dta file format. Attempting to write *Stata* dta files with strings\n",
    "longer than 244 characters raises a `ValueError`.\n",
    "\n",
    "\n",
    "<a id='io-stata-reader'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading from Stata format\n",
    "\n",
    "The top-level function `read_stata` will read a dta file and return\n",
    "either a `DataFrame` or a `StataReader` that can\n",
    "be used to read the file incrementally.\n",
    "\n",
    "Specifying a `chunksize` yields a\n",
    "`StataReader` instance that can be used to\n",
    "read `chunksize` lines from the file at a time.  The `StataReader`\n",
    "object can be used as an iterator.\n",
    "\n",
    "For more fine-grained control, use `iterator=True` and specify\n",
    "`chunksize` with each call to\n",
    "`read()`.\n",
    "\n",
    "Currently the `index` is retrieved as a column.\n",
    "\n",
    "The parameter `convert_categoricals` indicates whether value labels should be\n",
    "read and used to create a `Categorical` variable from them. Value labels can\n",
    "also be retrieved by the function `value_labels`, which requires `read()`\n",
    "to be called before use.\n",
    "\n",
    "The parameter `convert_missing` indicates whether missing value\n",
    "representations in Stata should be preserved.  If `False` (the default),\n",
    "missing values are represented as `np.nan`.  If `True`, missing values are\n",
    "represented using `StataMissingValue` objects, and columns containing missing\n",
    "values will have `object` data type.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">`read_stata()` and\n",
    "`StataReader` support .dta formats 113-115\n",
    "(Stata 10-12), 117 (Stata 13), and 118 (Stata 14).\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">Setting `preserve_dtypes=False` will upcast to the standard pandas data types:\n",
    "`int64` for all integer types and `float64` for floating point data.  By default,\n",
    "the Stata data types are preserved when importing.\n",
    "\n",
    "\n",
    "<a id='io-stata-categorical'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Categorical data\n",
    "\n",
    "`Categorical` data can be exported to *Stata* data files as value labeled data.\n",
    "The exported data consists of the underlying category codes as integer data values\n",
    "and the categories as value labels.  *Stata* does not have an explicit equivalent\n",
    "to a `Categorical` and information about *whether* the variable is ordered\n",
    "is lost when exporting.\n",
    "\n",
    "*Stata* only supports string value labels, and so `str` is called on the\n",
    "categories when exporting data.  Exporting `Categorical` variables with\n",
    "non-string categories produces a warning, and can result a loss of\n",
    "information if the `str` representations of the categories are not unique.\n",
    "\n",
    "Labeled data can similarly be imported from *Stata* data files as `Categorical`\n",
    "variables using the keyword argument `convert_categoricals` (`True` by default).\n",
    "The keyword argument `order_categoricals` (`True` by default) determines\n",
    "whether imported `Categorical` variables are ordered.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">When importing categorical data, the values of the variables in the *Stata*\n",
    "data file are not preserved since `Categorical` variables always\n",
    "use integer data types between `-1` and `n-1` where `n` is the number\n",
    "of categories. If the original values in the *Stata* data file are required,\n",
    "these can be imported by setting `convert_categoricals=False`, which will\n",
    "import original data (but not the variable labels). The original values can\n",
    "be matched to the imported categorical data since there is a simple mapping\n",
    "between the original *Stata* data values and the category codes of imported\n",
    "Categorical variables: missing values are assigned code `-1`, and the\n",
    "smallest original value is assigned `0`, the second smallest is assigned\n",
    "`1` and so on until the largest original value is assigned the code `n-1`.\n",
    "\n",
    ">**Note**\n",
    ">\n",
    ">*Stata* supports partially labeled series. These series have value labels for\n",
    "some but not all data values. Importing a partially labeled series will produce\n",
    "a `Categorical` with string categories for the values that are labeled and\n",
    "numeric categories for values with no label.\n",
    "\n",
    "\n",
    "<a id='io-sas'></a>\n",
    "\n",
    "<a id='io-sas-reader'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SAS formats\n",
    "\n",
    "The top-level function `read_sas()` can read (but not write) SAS\n",
    "xport (.XPT) and (since *v0.18.0*) SAS7BDAT (.sas7bdat) format files.\n",
    "\n",
    "SAS files only contain two value types: ASCII text and floating point\n",
    "values (usually 8 bytes but sometimes truncated).  For xport files,\n",
    "there is no automatic type conversion to integers, dates, or\n",
    "categoricals.  For SAS7BDAT files, the format codes may allow date\n",
    "variables to be automatically converted to dates.  By default the\n",
    "whole file is read and returned as a `DataFrame`.\n",
    "\n",
    "Specify a `chunksize` or use `iterator=True` to obtain reader\n",
    "objects (`XportReader` or `SAS7BDATReader`) for incrementally\n",
    "reading the file.  The reader objects also have attributes that\n",
    "contain additional information about the file and its variables.\n",
    "\n",
    "Read a SAS7BDAT file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "df = pd.read_sas('sas_data.sas7bdat')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Obtain an iterator and read an XPORT file 100,000 lines at a time:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "def do_something(chunk):\n",
    "    pass\n",
    "\n",
    "rdr = pd.read_sas('sas_xport.xpt', chunk=100000)\n",
    "for chunk in rdr:\n",
    "    do_something(chunk)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The [specification](https://support.sas.com/techsup/technote/ts140.pdf) for the xport file format is available from the SAS\n",
    "web site.\n",
    "\n",
    "No official documentation is available for the SAS7BDAT format.\n",
    "\n",
    "\n",
    "<a id='io-other'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Other file formats\n",
    "\n",
    "pandas itself only supports IO with a limited set of file formats that map\n",
    "cleanly to its tabular data model. For reading and writing other file formats\n",
    "into and from pandas, we recommend these packages from the broader community."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### netCDF\n",
    "\n",
    "[xarray](https://xarray.pydata.org/) provides data structures inspired by the pandas `DataFrame` for working\n",
    "with multi-dimensional datasets, with a focus on the netCDF file format and\n",
    "easy conversion to and from pandas.\n",
    "\n",
    "\n",
    "<a id='io-perf'></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Performance considerations\n",
    "\n",
    "This is an informal comparison of various IO methods, using pandas\n",
    "0.20.3. Timings are machine dependent and small differences should be\n",
    "ignored."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hide-output": false
   },
   "source": [
    "```ipython\n",
    "In [1]: sz = 1000000\n",
    "In [2]: df = pd.DataFrame({'A': np.random.randn(sz), 'B': [1] * sz})\n",
    "\n",
    "In [3]: df.info()\n",
    "<class 'pandas.core.frame.DataFrame'>\n",
    "RangeIndex: 1000000 entries, 0 to 999999\n",
    "Data columns (total 2 columns):\n",
    "A    1000000 non-null float64\n",
    "B    1000000 non-null int64\n",
    "dtypes: float64(1), int64(1)\n",
    "memory usage: 15.3 MB\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given the next test set:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hide-output": false
   },
   "outputs": [],
   "source": [
    "from numpy.random import randn\n",
    "\n",
    "sz = 1000000\n",
    "df = pd.DataFrame({'A': randn(sz), 'B': [1] * sz})\n",
    "\n",
    "\n",
    "def test_sql_write(df):\n",
    "    if os.path.exists('test.sql'):\n",
    "        os.remove('test.sql')\n",
    "    sql_db = sqlite3.connect('test.sql')\n",
    "    df.to_sql(name='test_table', con=sql_db)\n",
    "    sql_db.close()\n",
    "\n",
    "\n",
    "def test_sql_read():\n",
    "    sql_db = sqlite3.connect('test.sql')\n",
    "    pd.read_sql_query(\"select * from test_table\", sql_db)\n",
    "    sql_db.close()\n",
    "\n",
    "\n",
    "def test_hdf_fixed_write(df):\n",
    "    df.to_hdf('test_fixed.hdf', 'test', mode='w')\n",
    "\n",
    "\n",
    "def test_hdf_fixed_read():\n",
    "    pd.read_hdf('test_fixed.hdf', 'test')\n",
    "\n",
    "\n",
    "def test_hdf_fixed_write_compress(df):\n",
    "    df.to_hdf('test_fixed_compress.hdf', 'test', mode='w', complib='blosc')\n",
    "\n",
    "\n",
    "def test_hdf_fixed_read_compress():\n",
    "    pd.read_hdf('test_fixed_compress.hdf', 'test')\n",
    "\n",
    "\n",
    "def test_hdf_table_write(df):\n",
    "    df.to_hdf('test_table.hdf', 'test', mode='w', format='table')\n",
    "\n",
    "\n",
    "def test_hdf_table_read():\n",
    "    pd.read_hdf('test_table.hdf', 'test')\n",
    "\n",
    "\n",
    "def test_hdf_table_write_compress(df):\n",
    "    df.to_hdf('test_table_compress.hdf', 'test', mode='w',\n",
    "              complib='blosc', format='table')\n",
    "\n",
    "\n",
    "def test_hdf_table_read_compress():\n",
    "    pd.read_hdf('test_table_compress.hdf', 'test')\n",
    "\n",
    "\n",
    "def test_csv_write(df):\n",
    "    df.to_csv('test.csv', mode='w')\n",
    "\n",
    "\n",
    "def test_csv_read():\n",
    "    pd.read_csv('test.csv', index_col=0)\n",
    "\n",
    "\n",
    "def test_feather_write(df):\n",
    "    df.to_feather('test.feather')\n",
    "\n",
    "\n",
    "def test_feather_read():\n",
    "    pd.read_feather('test.feather')\n",
    "\n",
    "\n",
    "def test_pickle_write(df):\n",
    "    df.to_pickle('test.pkl')\n",
    "\n",
    "\n",
    "def test_pickle_read():\n",
    "    pd.read_pickle('test.pkl')\n",
    "\n",
    "\n",
    "def test_pickle_write_compress(df):\n",
    "    df.to_pickle('test.pkl.compress', compression='xz')\n",
    "\n",
    "\n",
    "def test_pickle_read_compress():\n",
    "    pd.read_pickle('test.pkl.compress', compression='xz')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When writing, the top-three functions in terms of speed are are\n",
    "`test_pickle_write`, `test_feather_write` and `test_hdf_fixed_write_compress`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hide-output": false
   },
   "source": [
    "```ipython\n",
    "In [14]: %timeit test_sql_write(df)\n",
    "2.37 s ± 36.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n",
    "\n",
    "In [15]: %timeit test_hdf_fixed_write(df)\n",
    "194 ms ± 65.9 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n",
    "\n",
    "In [26]: %timeit test_hdf_fixed_write_compress(df)\n",
    "119 ms ± 2.15 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n",
    "\n",
    "In [16]: %timeit test_hdf_table_write(df)\n",
    "623 ms ± 125 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n",
    "\n",
    "In [27]: %timeit test_hdf_table_write_compress(df)\n",
    "563 ms ± 23.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n",
    "\n",
    "In [17]: %timeit test_csv_write(df)\n",
    "3.13 s ± 49.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n",
    "\n",
    "In [30]: %timeit test_feather_write(df)\n",
    "103 ms ± 5.88 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n",
    "\n",
    "In [31]: %timeit test_pickle_write(df)\n",
    "109 ms ± 3.72 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n",
    "\n",
    "In [32]: %timeit test_pickle_write_compress(df)\n",
    "3.33 s ± 55.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When reading, the top three are `test_feather_read`, `test_pickle_read` and\n",
    "`test_hdf_fixed_read`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hide-output": false
   },
   "source": [
    "```ipython\n",
    "In [18]: %timeit test_sql_read()\n",
    "1.35 s ± 14.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n",
    "\n",
    "In [19]: %timeit test_hdf_fixed_read()\n",
    "14.3 ms ± 438 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)\n",
    "\n",
    "In [28]: %timeit test_hdf_fixed_read_compress()\n",
    "23.5 ms ± 672 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)\n",
    "\n",
    "In [20]: %timeit test_hdf_table_read()\n",
    "35.4 ms ± 314 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)\n",
    "\n",
    "In [29]: %timeit test_hdf_table_read_compress()\n",
    "42.6 ms ± 2.1 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n",
    "\n",
    "In [22]: %timeit test_csv_read()\n",
    "516 ms ± 27.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n",
    "\n",
    "In [33]: %timeit test_feather_read()\n",
    "4.06 ms ± 115 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)\n",
    "\n",
    "In [34]: %timeit test_pickle_read()\n",
    "6.5 ms ± 172 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)\n",
    "\n",
    "In [35]: %timeit test_pickle_read_compress()\n",
    "588 ms ± 3.57 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Space on disk (in bytes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hide-output": false
   },
   "source": [
    "```text\n",
    "34816000 Aug 21 18:00 test.sql\n",
    "24009240 Aug 21 18:00 test_fixed.hdf\n",
    " 7919610 Aug 21 18:00 test_fixed_compress.hdf\n",
    "24458892 Aug 21 18:00 test_table.hdf\n",
    " 8657116 Aug 21 18:00 test_table_compress.hdf\n",
    "28520770 Aug 21 18:00 test.csv\n",
    "16000248 Aug 21 18:00 test.feather\n",
    "16000848 Aug 21 18:00 test.pkl\n",
    " 7554108 Aug 21 18:00 test.pkl.compress\n",
    "```\n"
   ]
  }
 ],
 "metadata": {
  "filename": "io.rst",
  "kernelspec": {
   "display_name": "Python",
   "language": "python3",
   "name": "python3"
  },
  "title": "IO tools (text, CSV, HDF5, …)"
 },
 "nbformat": 4,
 "nbformat_minor": 2
}